{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import csv "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "PATH = \"./disgenet/\"\n",
    "filename = \"./disgenet/curated_gene_disease_associations.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(filename, sep = '\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(df), df.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#find row associated to disease Malignant Mesothelioma with id C0345967\n",
    "\n",
    "The columns in the files are:\n",
    "* geneId \t\t-> NCBI Entrez Gene Identifier\n",
    "* geneSymbol\t-> Official Gene Symbol\n",
    "* DSI\t\t-> The Disease Specificity Index for the gene\n",
    "* DPI\t\t-> The Disease Pleiotropy Index for the gene\n",
    "* diseaseId \t-> UMLS concept unique identifier\n",
    "* diseaseName \t-> Name of the disease\t\n",
    "* diseaseType  \t-> The DisGeNET disease type: disease, phenotype and group\n",
    "* diseaseClass\t-> The MeSH disease class(es)\n",
    "* diseaseSemanticType\t-> The UMLS Semantic Type(s) of the disease\n",
    "* score\t\t-> DisGENET score for the Gene-Disease association\n",
    "* EI\t\t-> The Evidence Index for the Gene-Disease association\n",
    "* YearInitial\t-> First time that the Gene-Disease association was reported\n",
    "* YearFinal\t-> Last time that the Gene-Disease association was reported\n",
    "* NofPmids\t-> Total number of publications reporting the Gene-Disease association\n",
    "* NofSnps\t\t-> Total number of SNPs associated to the Gene-Disease association\n",
    "* source\t\t-> Original source reporting the Gene-Disease association"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = df.loc[df['diseaseName'] == 'Malignant mesothelioma']\n",
    "target.to_csv(PATH+\"malignant_mesothelioma_curated_genes.tsv\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "t = pd.read_csv(PATH+\"malignant_mesothelioma_curated_genes.tsv\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneId</th>\n",
       "      <th>geneSymbol</th>\n",
       "      <th>DSI</th>\n",
       "      <th>DPI</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>diseaseName</th>\n",
       "      <th>diseaseType</th>\n",
       "      <th>diseaseClass</th>\n",
       "      <th>diseaseSemanticType</th>\n",
       "      <th>score</th>\n",
       "      <th>EI</th>\n",
       "      <th>YearInitial</th>\n",
       "      <th>YearFinal</th>\n",
       "      <th>NofPmids</th>\n",
       "      <th>NofSnps</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>0.466</td>\n",
       "      <td>0.828</td>\n",
       "      <td>C0345967</td>\n",
       "      <td>Malignant mesothelioma</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C08</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.5</td>\n",
       "      <td>1995.0</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>142</td>\n",
       "      <td>PARP1</td>\n",
       "      <td>0.432</td>\n",
       "      <td>0.862</td>\n",
       "      <td>C0345967</td>\n",
       "      <td>Malignant mesothelioma</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C08</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.32</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>302</td>\n",
       "      <td>ANXA2</td>\n",
       "      <td>0.485</td>\n",
       "      <td>0.793</td>\n",
       "      <td>C0345967</td>\n",
       "      <td>Malignant mesothelioma</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C08</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>335</td>\n",
       "      <td>APOA1</td>\n",
       "      <td>0.463</td>\n",
       "      <td>0.759</td>\n",
       "      <td>C0345967</td>\n",
       "      <td>Malignant mesothelioma</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C08</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>596</td>\n",
       "      <td>BCL2</td>\n",
       "      <td>0.312</td>\n",
       "      <td>0.862</td>\n",
       "      <td>C0345967</td>\n",
       "      <td>Malignant mesothelioma</td>\n",
       "      <td>disease</td>\n",
       "      <td>C04;C08</td>\n",
       "      <td>Neoplastic Process</td>\n",
       "      <td>0.31</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>CTD_human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   geneId geneSymbol    DSI    DPI diseaseId             diseaseName  \\\n",
       "0      10       NAT2  0.466  0.828  C0345967  Malignant mesothelioma   \n",
       "1     142      PARP1  0.432  0.862  C0345967  Malignant mesothelioma   \n",
       "2     302      ANXA2  0.485  0.793  C0345967  Malignant mesothelioma   \n",
       "3     335      APOA1  0.463  0.759  C0345967  Malignant mesothelioma   \n",
       "4     596       BCL2  0.312  0.862  C0345967  Malignant mesothelioma   \n",
       "\n",
       "  diseaseType diseaseClass diseaseSemanticType  score   EI  YearInitial  \\\n",
       "0     disease      C04;C08  Neoplastic Process   0.34  0.5       1995.0   \n",
       "1     disease      C04;C08  Neoplastic Process   0.32  1.0       2011.0   \n",
       "2     disease      C04;C08  Neoplastic Process   0.30  NaN       2010.0   \n",
       "3     disease      C04;C08  Neoplastic Process   0.30  NaN       2013.0   \n",
       "4     disease      C04;C08  Neoplastic Process   0.31  1.0       2006.0   \n",
       "\n",
       "   YearFinal  NofPmids  NofSnps     source  \n",
       "0     2009.0         1        0  CTD_human  \n",
       "1     2011.0         1        0  CTD_human  \n",
       "2     2010.0         1        0  CTD_human  \n",
       "3     2013.0         1        0  CTD_human  \n",
       "4     2010.0         1        0  CTD_human  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = t.drop('Unnamed: 0', axis = 1)\n",
    "t.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Explore the DisGeNet dataset, find the disease of interest and get the list of human genes involved."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_id</th>\n",
       "      <th>UniProt</th>\n",
       "      <th>Gene_Full_Name</th>\n",
       "      <th>Protein_Class</th>\n",
       "      <th>pLI</th>\n",
       "      <th>DSI_g</th>\n",
       "      <th>DPI_g</th>\n",
       "      <th>N_diseases</th>\n",
       "      <th>N_SNPs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NAT2</td>\n",
       "      <td>10</td>\n",
       "      <td>P11245</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>transferase</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.466</td>\n",
       "      <td>0.828</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PARP1</td>\n",
       "      <td>142</td>\n",
       "      <td>P09874</td>\n",
       "      <td>poly(ADP-ribose) polymerase 1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000552</td>\n",
       "      <td>0.432</td>\n",
       "      <td>0.862</td>\n",
       "      <td>66</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ANXA2</td>\n",
       "      <td>302</td>\n",
       "      <td>P07355</td>\n",
       "      <td>annexin A2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.485</td>\n",
       "      <td>0.793</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>APOA1</td>\n",
       "      <td>335</td>\n",
       "      <td>P02647</td>\n",
       "      <td>apolipoprotein A1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000549</td>\n",
       "      <td>0.463</td>\n",
       "      <td>0.759</td>\n",
       "      <td>54</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BCL2</td>\n",
       "      <td>596</td>\n",
       "      <td>P10415</td>\n",
       "      <td>BCL2, apoptosis regulator</td>\n",
       "      <td>signaling molecule</td>\n",
       "      <td>0.567050</td>\n",
       "      <td>0.312</td>\n",
       "      <td>0.862</td>\n",
       "      <td>137</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene  Gene_id UniProt                 Gene_Full_Name       Protein_Class  \\\n",
       "0   NAT2       10  P11245          N-acetyltransferase 2         transferase   \n",
       "1  PARP1      142  P09874  poly(ADP-ribose) polymerase 1                 NaN   \n",
       "2  ANXA2      302  P07355                     annexin A2                 NaN   \n",
       "3  APOA1      335  P02647              apolipoprotein A1                 NaN   \n",
       "4   BCL2      596  P10415      BCL2, apoptosis regulator  signaling molecule   \n",
       "\n",
       "        pLI  DSI_g  DPI_g  N_diseases  N_SNPs  \n",
       "0  0.000004  0.466  0.828          37       1  \n",
       "1  0.000552  0.432  0.862          66       2  \n",
       "2  0.000482  0.485  0.793          26       0  \n",
       "3  0.000549  0.463  0.759          54      15  \n",
       "4  0.567050  0.312  0.862         137       8  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "curated = pd.read_csv(\"./disgenet/browser_source_genes_summary_CURATED.tsv\", sep = '\\t')\n",
    "curated.head()            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#now let's save the gene symbols, entrez ID and names in arrays\n",
    "geneSymbol=[]\n",
    "geneID=[]\n",
    "geneName=[]\n",
    "for i in range(len(curated)):\n",
    "    geneSymbol.append(curated['Gene'][i])\n",
    "    geneID.append(curated['Gene_id'][i])\n",
    "    geneName.append(curated['Gene_Full_Name'][i])    \n",
    "#we check on HGNC to see if we need to change genes name"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "b) For all genes in the seed gene list, collect the following basic information from the Uniprot:\n",
    "\n",
    "* official (primary) gene symbol (check if the symbols are updated and approved on the HGNC website; report any issue/lack of data/potential misinterpretation)\n",
    "\n",
    "* Uniprot AC, alphanumeric ‘accession number’ (a.k.a. ’Uniprot entry’)\n",
    "\n",
    "* protein name (the main one only, do not report the aliases)\n",
    "\n",
    "* Entrez Gene ID (a.k.a. ‘GeneID’) very brief description of its function (keep it very short, i.e. max 20 words)\n",
    "\n",
    "* notes related to the above information, if any and if relevant\n",
    "\n",
    "Store the data gathered in a table in an easily accessible format of your choice (csv, tab,\n",
    "excel, etc)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(geneSymbol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fino a COL12A1 approved"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# printing original list  \n",
    "print(\"The original list is : \" + str(geneSymbol))\n",
    "  \n",
    "# using join() \n",
    "# avoiding printing last comma \n",
    "print(\"The formatted output is : \") \n",
    "print(', '.join(geneSymbol)) \n",
    "\n",
    "# deleted quotes to pass the list into site\n",
    "#https://www.genenames.org/tools/multi-symbol-checker/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# all gene symbols were approved by HGNC "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprotAC=[]\n",
    "for i in range(len(curated)):\n",
    "    uniprotAC.append(curated['UniProt'][i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results = pd.DataFrame(list(zip(geneSymbol,geneName,geneID,uniprotAC)), columns=['Symbol','Name','ID','UniprotAC'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results.to_csv('mesothelioma-curated-genes.csv')\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercise 1.2\n",
    "\n",
    "For each seed gene, collect all binary protein interactions from two different PPI sources:\n",
    "* Biogrid Human, latest release available\n",
    "* IID Integrated Interactions Database (experimental data only, all tissues, unless stated otherwise in further instruction)\n",
    "\n",
    "Note: once you got the list of the proteins interacting with at least one seed gene, you must\n",
    "also retrieve and include in your interactome the interactions among these non-seed\n",
    "proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/giulia/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (1,2,18) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "#open biogrid DB\n",
    "biogrid=pd.read_csv('BIOGRID-ALL-3.5.179.tab2.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#BioGRID Interaction ID</th>\n",
       "      <th>Entrez Gene Interactor A</th>\n",
       "      <th>Entrez Gene Interactor B</th>\n",
       "      <th>BioGRID ID Interactor A</th>\n",
       "      <th>BioGRID ID Interactor B</th>\n",
       "      <th>Systematic Name Interactor A</th>\n",
       "      <th>Systematic Name Interactor B</th>\n",
       "      <th>Official Symbol Interactor A</th>\n",
       "      <th>Official Symbol Interactor B</th>\n",
       "      <th>Synonyms Interactor A</th>\n",
       "      <th>...</th>\n",
       "      <th>Pubmed ID</th>\n",
       "      <th>Organism Interactor A</th>\n",
       "      <th>Organism Interactor B</th>\n",
       "      <th>Throughput</th>\n",
       "      <th>Score</th>\n",
       "      <th>Modification</th>\n",
       "      <th>Phenotypes</th>\n",
       "      <th>Qualifications</th>\n",
       "      <th>Tags</th>\n",
       "      <th>Source Database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>103</td>\n",
       "      <td>6416</td>\n",
       "      <td>2318</td>\n",
       "      <td>112315</td>\n",
       "      <td>108607</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MAP2K4</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...</td>\n",
       "      <td>...</td>\n",
       "      <td>9006895</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>117</td>\n",
       "      <td>84665</td>\n",
       "      <td>88</td>\n",
       "      <td>124185</td>\n",
       "      <td>106603</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MYPN</td>\n",
       "      <td>ACTN2</td>\n",
       "      <td>CMD1DD|CMH22|MYOP|RCM4</td>\n",
       "      <td>...</td>\n",
       "      <td>11309420</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>183</td>\n",
       "      <td>90</td>\n",
       "      <td>2339</td>\n",
       "      <td>106605</td>\n",
       "      <td>108625</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>FNTA</td>\n",
       "      <td>ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI</td>\n",
       "      <td>...</td>\n",
       "      <td>8599089</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>278</td>\n",
       "      <td>2624</td>\n",
       "      <td>5371</td>\n",
       "      <td>108894</td>\n",
       "      <td>111384</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "      <td>DCML|IMD21|MONOMAC|NFE1B</td>\n",
       "      <td>...</td>\n",
       "      <td>10938104</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>418</td>\n",
       "      <td>6118</td>\n",
       "      <td>6774</td>\n",
       "      <td>112038</td>\n",
       "      <td>112651</td>\n",
       "      <td>RP4-547C9.3</td>\n",
       "      <td>-</td>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>REPA2|RP-A p32|RP-A p34|RPA32</td>\n",
       "      <td>...</td>\n",
       "      <td>10875894</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #BioGRID Interaction ID Entrez Gene Interactor A Entrez Gene Interactor B  \\\n",
       "0                      103                     6416                     2318   \n",
       "1                      117                    84665                       88   \n",
       "2                      183                       90                     2339   \n",
       "3                      278                     2624                     5371   \n",
       "4                      418                     6118                     6774   \n",
       "\n",
       "   BioGRID ID Interactor A  BioGRID ID Interactor B  \\\n",
       "0                   112315                   108607   \n",
       "1                   124185                   106603   \n",
       "2                   106605                   108625   \n",
       "3                   108894                   111384   \n",
       "4                   112038                   112651   \n",
       "\n",
       "  Systematic Name Interactor A Systematic Name Interactor B  \\\n",
       "0                            -                            -   \n",
       "1                            -                            -   \n",
       "2                            -                            -   \n",
       "3                            -                            -   \n",
       "4                  RP4-547C9.3                            -   \n",
       "\n",
       "  Official Symbol Interactor A Official Symbol Interactor B  \\\n",
       "0                       MAP2K4                         FLNC   \n",
       "1                         MYPN                        ACTN2   \n",
       "2                        ACVR1                         FNTA   \n",
       "3                        GATA2                          PML   \n",
       "4                         RPA2                        STAT3   \n",
       "\n",
       "                               Synonyms Interactor A  ... Pubmed ID  \\\n",
       "0  JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...  ...   9006895   \n",
       "1                             CMD1DD|CMH22|MYOP|RCM4  ...  11309420   \n",
       "2            ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI  ...   8599089   \n",
       "3                           DCML|IMD21|MONOMAC|NFE1B  ...  10938104   \n",
       "4                      REPA2|RP-A p32|RP-A p34|RPA32  ...  10875894   \n",
       "\n",
       "  Organism Interactor A Organism Interactor B      Throughput  Score  \\\n",
       "0                  9606                  9606  Low Throughput      -   \n",
       "1                  9606                  9606  Low Throughput      -   \n",
       "2                  9606                  9606  Low Throughput      -   \n",
       "3                  9606                  9606  Low Throughput      -   \n",
       "4                  9606                  9606  Low Throughput      -   \n",
       "\n",
       "   Modification  Phenotypes Qualifications Tags Source Database  \n",
       "0             -           -              -    -         BIOGRID  \n",
       "1             -           -              -    -         BIOGRID  \n",
       "2             -           -              -    -         BIOGRID  \n",
       "3             -           -              -    -         BIOGRID  \n",
       "4             -           -              -    -         BIOGRID  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biogrid.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1730436"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(biogrid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['#BioGRID Interaction ID', 'Entrez Gene Interactor A',\n",
       "       'Entrez Gene Interactor B', 'BioGRID ID Interactor A',\n",
       "       'BioGRID ID Interactor B', 'Systematic Name Interactor A',\n",
       "       'Systematic Name Interactor B', 'Official Symbol Interactor A',\n",
       "       'Official Symbol Interactor B', 'Synonyms Interactor A',\n",
       "       'Synonyms Interactor B', 'Experimental System',\n",
       "       'Experimental System Type', 'Author', 'Pubmed ID',\n",
       "       'Organism Interactor A', 'Organism Interactor B', 'Throughput', 'Score',\n",
       "       'Modification', 'Phenotypes', 'Qualifications', 'Tags',\n",
       "       'Source Database'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#let's select what we want from the biogrid dataset.\n",
    "#select only human PPI\n",
    "biogrid.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#I took the interactions for NAT2 gene in humans \n",
    "example = pd.read_csv('./biogrid/BIOGRID-GENE-106528-3.5.179.tab2.txt', sep='\\t')\n",
    "example.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#BioGRID Interaction ID</th>\n",
       "      <th>Entrez Gene Interactor A</th>\n",
       "      <th>Entrez Gene Interactor B</th>\n",
       "      <th>BioGRID ID Interactor A</th>\n",
       "      <th>BioGRID ID Interactor B</th>\n",
       "      <th>Systematic Name Interactor A</th>\n",
       "      <th>Systematic Name Interactor B</th>\n",
       "      <th>Official Symbol Interactor A</th>\n",
       "      <th>Official Symbol Interactor B</th>\n",
       "      <th>Synonyms Interactor A</th>\n",
       "      <th>...</th>\n",
       "      <th>Pubmed ID</th>\n",
       "      <th>Organism Interactor A</th>\n",
       "      <th>Organism Interactor B</th>\n",
       "      <th>Throughput</th>\n",
       "      <th>Score</th>\n",
       "      <th>Modification</th>\n",
       "      <th>Phenotypes</th>\n",
       "      <th>Qualifications</th>\n",
       "      <th>Tags</th>\n",
       "      <th>Source Database</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>103</td>\n",
       "      <td>6416</td>\n",
       "      <td>2318</td>\n",
       "      <td>112315</td>\n",
       "      <td>108607</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MAP2K4</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...</td>\n",
       "      <td>...</td>\n",
       "      <td>9006895</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>117</td>\n",
       "      <td>84665</td>\n",
       "      <td>88</td>\n",
       "      <td>124185</td>\n",
       "      <td>106603</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MYPN</td>\n",
       "      <td>ACTN2</td>\n",
       "      <td>CMD1DD|CMH22|MYOP|RCM4</td>\n",
       "      <td>...</td>\n",
       "      <td>11309420</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>183</td>\n",
       "      <td>90</td>\n",
       "      <td>2339</td>\n",
       "      <td>106605</td>\n",
       "      <td>108625</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>FNTA</td>\n",
       "      <td>ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI</td>\n",
       "      <td>...</td>\n",
       "      <td>8599089</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>278</td>\n",
       "      <td>2624</td>\n",
       "      <td>5371</td>\n",
       "      <td>108894</td>\n",
       "      <td>111384</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "      <td>DCML|IMD21|MONOMAC|NFE1B</td>\n",
       "      <td>...</td>\n",
       "      <td>10938104</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>418</td>\n",
       "      <td>6118</td>\n",
       "      <td>6774</td>\n",
       "      <td>112038</td>\n",
       "      <td>112651</td>\n",
       "      <td>RP4-547C9.3</td>\n",
       "      <td>-</td>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>REPA2|RP-A p32|RP-A p34|RPA32</td>\n",
       "      <td>...</td>\n",
       "      <td>10875894</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1730424</th>\n",
       "      <td>2621674</td>\n",
       "      <td>84376</td>\n",
       "      <td>10749</td>\n",
       "      <td>124068</td>\n",
       "      <td>115972</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>HOOK3</td>\n",
       "      <td>KIF1C</td>\n",
       "      <td>HK3</td>\n",
       "      <td>...</td>\n",
       "      <td>28718761</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1730425</th>\n",
       "      <td>2621675</td>\n",
       "      <td>2059</td>\n",
       "      <td>50618</td>\n",
       "      <td>108373</td>\n",
       "      <td>119098</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>EPS8</td>\n",
       "      <td>ITSN2</td>\n",
       "      <td>DFNB102</td>\n",
       "      <td>...</td>\n",
       "      <td>22449706</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1730426</th>\n",
       "      <td>2621676</td>\n",
       "      <td>2059</td>\n",
       "      <td>50618</td>\n",
       "      <td>108373</td>\n",
       "      <td>119098</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>EPS8</td>\n",
       "      <td>ITSN2</td>\n",
       "      <td>DFNB102</td>\n",
       "      <td>...</td>\n",
       "      <td>22449706</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1730427</th>\n",
       "      <td>2621677</td>\n",
       "      <td>50618</td>\n",
       "      <td>2059</td>\n",
       "      <td>119098</td>\n",
       "      <td>108373</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ITSN2</td>\n",
       "      <td>EPS8</td>\n",
       "      <td>PRO2015|SH3D1B|SH3P18|SWA|SWAP</td>\n",
       "      <td>...</td>\n",
       "      <td>22449706</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1730428</th>\n",
       "      <td>2621678</td>\n",
       "      <td>2059</td>\n",
       "      <td>50618</td>\n",
       "      <td>108373</td>\n",
       "      <td>119098</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>EPS8</td>\n",
       "      <td>ITSN2</td>\n",
       "      <td>DFNB102</td>\n",
       "      <td>...</td>\n",
       "      <td>22449706</td>\n",
       "      <td>9606</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>485380 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #BioGRID Interaction ID Entrez Gene Interactor A  \\\n",
       "0                            103                     6416   \n",
       "1                            117                    84665   \n",
       "2                            183                       90   \n",
       "3                            278                     2624   \n",
       "4                            418                     6118   \n",
       "...                          ...                      ...   \n",
       "1730424                  2621674                    84376   \n",
       "1730425                  2621675                     2059   \n",
       "1730426                  2621676                     2059   \n",
       "1730427                  2621677                    50618   \n",
       "1730428                  2621678                     2059   \n",
       "\n",
       "        Entrez Gene Interactor B  BioGRID ID Interactor A  \\\n",
       "0                           2318                   112315   \n",
       "1                             88                   124185   \n",
       "2                           2339                   106605   \n",
       "3                           5371                   108894   \n",
       "4                           6774                   112038   \n",
       "...                          ...                      ...   \n",
       "1730424                    10749                   124068   \n",
       "1730425                    50618                   108373   \n",
       "1730426                    50618                   108373   \n",
       "1730427                     2059                   119098   \n",
       "1730428                    50618                   108373   \n",
       "\n",
       "         BioGRID ID Interactor B Systematic Name Interactor A  \\\n",
       "0                         108607                            -   \n",
       "1                         106603                            -   \n",
       "2                         108625                            -   \n",
       "3                         111384                            -   \n",
       "4                         112651                  RP4-547C9.3   \n",
       "...                          ...                          ...   \n",
       "1730424                   115972                            -   \n",
       "1730425                   119098                            -   \n",
       "1730426                   119098                            -   \n",
       "1730427                   108373                            -   \n",
       "1730428                   119098                            -   \n",
       "\n",
       "        Systematic Name Interactor B Official Symbol Interactor A  \\\n",
       "0                                  -                       MAP2K4   \n",
       "1                                  -                         MYPN   \n",
       "2                                  -                        ACVR1   \n",
       "3                                  -                        GATA2   \n",
       "4                                  -                         RPA2   \n",
       "...                              ...                          ...   \n",
       "1730424                            -                        HOOK3   \n",
       "1730425                            -                         EPS8   \n",
       "1730426                            -                         EPS8   \n",
       "1730427                            -                        ITSN2   \n",
       "1730428                            -                         EPS8   \n",
       "\n",
       "        Official Symbol Interactor B  \\\n",
       "0                               FLNC   \n",
       "1                              ACTN2   \n",
       "2                               FNTA   \n",
       "3                                PML   \n",
       "4                              STAT3   \n",
       "...                              ...   \n",
       "1730424                        KIF1C   \n",
       "1730425                        ITSN2   \n",
       "1730426                        ITSN2   \n",
       "1730427                         EPS8   \n",
       "1730428                        ITSN2   \n",
       "\n",
       "                                     Synonyms Interactor A  ... Pubmed ID  \\\n",
       "0        JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...  ...   9006895   \n",
       "1                                   CMD1DD|CMH22|MYOP|RCM4  ...  11309420   \n",
       "2                  ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI  ...   8599089   \n",
       "3                                 DCML|IMD21|MONOMAC|NFE1B  ...  10938104   \n",
       "4                            REPA2|RP-A p32|RP-A p34|RPA32  ...  10875894   \n",
       "...                                                    ...  ...       ...   \n",
       "1730424                                                HK3  ...  28718761   \n",
       "1730425                                            DFNB102  ...  22449706   \n",
       "1730426                                            DFNB102  ...  22449706   \n",
       "1730427                     PRO2015|SH3D1B|SH3P18|SWA|SWAP  ...  22449706   \n",
       "1730428                                            DFNB102  ...  22449706   \n",
       "\n",
       "        Organism Interactor A Organism Interactor B      Throughput  Score  \\\n",
       "0                        9606                  9606  Low Throughput      -   \n",
       "1                        9606                  9606  Low Throughput      -   \n",
       "2                        9606                  9606  Low Throughput      -   \n",
       "3                        9606                  9606  Low Throughput      -   \n",
       "4                        9606                  9606  Low Throughput      -   \n",
       "...                       ...                   ...             ...    ...   \n",
       "1730424                  9606                  9606  Low Throughput      -   \n",
       "1730425                  9606                  9606  Low Throughput      -   \n",
       "1730426                  9606                  9606  Low Throughput      -   \n",
       "1730427                  9606                  9606  Low Throughput      -   \n",
       "1730428                  9606                  9606  Low Throughput      -   \n",
       "\n",
       "         Modification  Phenotypes Qualifications Tags Source Database  \n",
       "0                   -           -              -    -         BIOGRID  \n",
       "1                   -           -              -    -         BIOGRID  \n",
       "2                   -           -              -    -         BIOGRID  \n",
       "3                   -           -              -    -         BIOGRID  \n",
       "4                   -           -              -    -         BIOGRID  \n",
       "...               ...         ...            ...  ...             ...  \n",
       "1730424             -           -              -    -         BIOGRID  \n",
       "1730425             -           -              -    -         BIOGRID  \n",
       "1730426             -           -              -    -         BIOGRID  \n",
       "1730427             -           -              -    -         BIOGRID  \n",
       "1730428             -           -              -    -         BIOGRID  \n",
       "\n",
       "[485380 rows x 24 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#select only human genes\n",
    "biogrid_human=biogrid.loc[(biogrid['Organism Interactor A']==9606) & (biogrid['Organism Interactor B']==9606)]\n",
    "biogrid_human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# look for the genes which interacts with at least one seed genes\n",
    "biogrid_seed_genes = biogrid_human.loc[(biogrid_human['Official Symbol Interactor A'].isin(geneSymbol)) | (biogrid_human['Official Symbol Interactor B'].isin(geneSymbol))]\n",
    "biogrid_seed_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## from our seed interaction let's search non-seeds genes that interacts with at least one seed gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Interactor A is not a seed gene but Interactor B is a seed gene\n",
    "non_seed_df_A = biogrid_seed_genes.loc[~(biogrid_seed_genes['Official Symbol Interactor A'].isin(geneSymbol))]\n",
    "seed_B = non_seed_df_A.loc[(non_seed_df_A ['Official Symbol Interactor B'].isin(geneSymbol))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# interactor B is not a seed genes but Interactor A is a seed gene\n",
    "non_seed_df_B = biogrid_seed_genes.loc[~(biogrid_seed_genes['Official Symbol Interactor B'].isin(geneSymbol))]\n",
    "non_seed_df_B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reset Index, otherwise they don't work with list\n",
    "non_seed_df_A=non_seed_df_A.reset_index(drop=True)\n",
    "non_seed_df_B=non_seed_df_B.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_seed_df_B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_seed_df_A = biogrid_seed_genes.loc[~(biogrid_seed_genes['Official Symbol Interactor A'].isin(geneSymbol))]\n",
    "non_seed_df_B = biogrid_seed_genes.loc[~(biogrid_seed_genes['Official Symbol Interactor B'].isin(geneSymbol))]\n",
    "non_seed_df_A=non_seed_df_A.reset_index(drop=True)\n",
    "non_seed_df_B=non_seed_df_B.reset_index(drop=True)\n",
    "#build a list with only non seed genes that interacts with at least one seed gene\n",
    "non_seed_list=[]\n",
    "for i in range(0, len(non_seed_df_A)):\n",
    "    non_seed_list.append(non_seed_df_A['Official Symbol Interactor A'][i])\n",
    "for i in range(0, len(non_seed_df_B)):\n",
    "    non_seed_list.append(non_seed_df_B['Official Symbol Interactor B'][i])\n",
    "#drop duplicates\n",
    "non_seed_list = list(dict.fromkeys(non_seed_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(0, len(non_seed_df_B)):\n",
    "    non_seed_list.append(non_seed_df_B['Official Symbol Interactor B'][i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#drop duplicates\n",
    "non_seed_list = list(dict.fromkeys(non_seed_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Verify Correctness\n",
    "for i in range(0, len(geneSymbol)):\n",
    "    for j in range(0, len(non_seed_list)):\n",
    "        if(geneSymbol[i] == non_seed_list[j]):\n",
    "            print(\"Something is wrong\")\n",
    "print(\"Ok\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(non_seed_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#now search for non seed interactions from the human DB\n",
    "#Return positions of non-seed genes that interacts with a non seed gene but both interacts with at least one seed gene\n",
    "biogrid_human = biogrid_human.reset_index(drop=True)\n",
    "#create a list of index of the original matrix with non seed genes interacting each others\n",
    "list_of_idx=[]\n",
    "#fill the list\n",
    "for i in range(len(biogrid_human)):\n",
    "    if biogrid_human['Official Symbol Interactor A'][i] in non_seed_list and biogrid_human['Official Symbol Interactor B'][i] in non_seed_list:\n",
    "        list_of_idx.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#biogrid_seed_genes\n",
    "biogrid_non_seed = biogrid_human.loc[list_of_idx]\n",
    "biogrid_non_seed = biogrid_non_seed.reset_index(drop=True)\n",
    "biogrid_non_seed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make a table in which interactor 1 is seed and interactor 2 can be seed or non-seed\n",
    "biogrid_seed_df = biogrid_seed_genes.loc[(biogrid_seed_genes['Official Symbol Interactor A'].isin(geneSymbol))]\n",
    "interactome = pd.concat([biogrid_seed_df, biogrid_non_seed], ignore_index=True)\n",
    "interactome.to_csv(\"interactome-biogrid.txt\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interactome = pd.read_csv(\"interactome-biogrid.txt\", sep= '\\t')\n",
    "interactome = interactome.drop(['Unnamed: 0'], axis = 1)\n",
    "interactome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iid = pd.read_csv('human_annotated_PPIs.txt', sep='\\t')\n",
    "iid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#select only rows with evidence type exp, drop some columns.\n",
    "iid = iid.loc[iid['evidence type'] == 'exp']\n",
    "icols = ['uniprot1', 'uniprot2', 'symbol1', 'symbol2' , 'evidence type', 'cancer']\n",
    "iid = iid.reset_index(drop=True)\n",
    "iid = iid[['uniprot1', 'uniprot2', 'symbol1', 'symbol2' , 'evidence type', 'cancer']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iid.to_csv('iid.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SAME PROCEDURE. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iid_seed_genes = iid.loc[(iid['symbol1'].isin(geneSymbol)) | (iid['symbol2'].isin(geneSymbol))]\n",
    "iid_seed_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_seed_1 = iid_seed_genes.loc[~(iid_seed_genes['symbol1'].isin(geneSymbol))]\n",
    "# interactor 1 is not a seed genes and interacts with a seed gene\n",
    "#B is a seed gene\n",
    "non_seed_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_seed_2 = iid_seed_genes.loc[~(iid_seed_genes['symbol2'].isin(geneSymbol))]\n",
    "# interactor 1 is not a seed genes and interacts with a seed gene\n",
    "#B is a seed gene\n",
    "non_seed_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_seed_1=non_seed_1.reset_index(drop=True)\n",
    "non_seed_2=non_seed_2.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#build a list with only non seed genes that interacts with at least one seed gene\n",
    "non_seed=[]\n",
    "#non_seed_df_A = biogrid_seed_genes.loc[~(biogrid_seed_genes['Official Symbol Interactor A'].isin(geneSymbol))]\n",
    "for i in range(0, len(non_seed_1)):\n",
    "    non_seed.append(non_seed_1['symbol1'][i])\n",
    "for i in range(0, len(non_seed_2)):\n",
    "    non_seed.append(non_seed_2['symbol2'][i])\n",
    "#drop duplicates\n",
    "non_seed = list(dict.fromkeys(non_seed))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(non_seed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iid=iid.reset_index(drop=True)\n",
    "#non_seed_df_A = biogrid_seed_genes.loc[~(biogrid_seed_genes['Official Symbol Interactor A'].isin(geneSymbol))]\n",
    "list_of_idx=[]\n",
    "#fill the list\n",
    "for i in range(len(iid_seed_genes)):\n",
    "    if iid['symbol1'][i] in non_seed and iid['symbol2'][i] in non_seed:\n",
    "        list_of_idx.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#biogrid_seed_genes\n",
    "iid_non_seed = iid.loc[list_of_idx]\n",
    "iid_non_seed = iid_non_seed.reset_index(drop=True)\n",
    "iid_non_seed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iid_seed_df = iid_seed_genes.loc[(iid_seed_genes['symbol1'].isin(geneSymbol))]\n",
    "interactome2 = pd.concat([iid_seed_df, iid_non_seed], ignore_index=True)\n",
    "interactome2.to_csv(\"interactome-iid.txt\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interactome2 = pd.read_csv(\"interactome-iid.txt\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interactome2 = interactome2.drop(['Unnamed: 0'], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interactome2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interactome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Summarize the main results in a table reporting:\n",
    "* no. of seed genes found in each different DBs (some seed genes may be missing in the DBs);\n",
    "* total no. of interacting proteins, including seed genes, for each DB;\n",
    "* total no. of interactions found in each DB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(geneSymbol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seed_B = non_seed_df_A.loc[(non_seed_df_A ['Official Symbol Interactor B'].isin(geneSymbol))]\n",
    "seed_A = non_seed_df_B.loc[(non_seed_df_B ['Official Symbol Interactor A'].isin(geneSymbol))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_unique_genes(dataframe, column_name):\n",
    "    genes_found = []\n",
    "    for index in range(len(dataframe)):\n",
    "        if dataframe[column_name][index] not in genes_found:\n",
    "            genes_found.append(dataframe[column_name][index])\n",
    "    return genes_found"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genes from Biogrid\n",
    "la = find_unique_genes(seed_B, 'Official Symbol Interactor B')\n",
    "lb = find_unique_genes(seed_A, 'Official Symbol Interactor A')\n",
    "l_tot = la + lb\n",
    "#drop duplicates\n",
    "total_genes = list(dict.fromkeys(l_tot))\n",
    "len(total_genes)\n",
    "missing_gene = []\n",
    "for index in range(len(geneSymbol)):\n",
    "    if geneSymbol[index] not in total_genes:\n",
    "        missing_gene.append(geneSymbol[index])\n",
    "print(\"Genes missing in Biogrid\\n\", missing_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Genes from IID \n",
    "seed1 = non_seed_2.loc[(non_seed_2['symbol1'].isin(geneSymbol))]\n",
    "seed2 = non_seed_1.loc[(non_seed_1['symbol2'].isin(geneSymbol))]\n",
    "# interactor 1 is not a seed genes and interacts with a seed gene\n",
    "#B is a seed gene\n",
    "l1 = find_unique_genes(seed1, 'symbol1')\n",
    "l2 = find_unique_genes(seed2, 'symbol2')\n",
    "l = l1+l2\n",
    "#drop duplicates\n",
    "i_genes = list(dict.fromkeys(l))\n",
    "iid_missing = []\n",
    "for index in range(len(geneSymbol)):\n",
    "    if geneSymbol[index] not in i_genes:\n",
    "        iid_missing.append(geneSymbol[index])\n",
    "print(\"Genes missing from IID DB \\n\", iid_missing)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "total no. of interacting proteins, including seed genes, for each DB;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interactors= []\n",
    "for index in range(len(biogrid_human)):\n",
    "    if biogrid_human['Official Symbol Interactor A'][index] not in interactors:\n",
    "        interactors.append(biogrid_human['Official Symbol Interactor A'][index])\n",
    "    if biogrid_human['Official Symbol Interactor B'][index] not in interactors:\n",
    "        interactors.append(biogrid_human['Official Symbol Interactor B'][index])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(interactors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_interactors(dataframe, column1, column2):\n",
    "    interactors= []\n",
    "    for index in range(len(dataframe)):\n",
    "        if dataframe[column1][index] not in interactors:\n",
    "            interactors.append(dataframe[column1][index])\n",
    "        if dataframe[column2][index] not in interactors:\n",
    "            interactors.append(dataframe[column2][index])\n",
    "    return len(interactors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "iid_len = count_interactors(iid, 'symbol1', 'symbol2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(iid_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(iid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# total no. of interactions found in each DB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "countbio = biogrid_human['#BioGRID Interaction ID'].nunique()\n",
    "print(countbio)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uniquesyms = []\n",
    "for index in range(0, len(iid)):\n",
    "    sym1 = iid['symbol1'][index]\n",
    "    sym2 = iid['symbol2'][index]\n",
    "    uniquesyms.append(sym1+sym2)\n",
    "#drop duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uniquesyms = list(dict.fromkeys(uniquesyms))\n",
    "print(len(uniquesyms))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build and store three tables:\n",
    "\n",
    "* seed genes interactome: interactions that involve seed genes only, from all DBs, in the format:\n",
    "  interactor A gene symbol, interactor B gene symbol, interactor A Uniprot AC, interactor B\n",
    "  Uniprot AC, database source\n",
    "* union interactome: all proteins interacting with at least one seed gene, from all DBs, same format as above.\n",
    "* intersection interactome: all proteins interacting with at least one seed gene confirmed by both DBs, in the       format: interactor A gene symbol, interactor B gene symbol, interactor A Uniprot AC, interactor B Uniprot AC\n",
    "\n",
    "Always check that interactors are both human (i.e. organism ID is always 9606, Homo\n",
    "Sapiens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "biogrid_human = biogrid_human.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_human = pd.read_csv(\"./uniprot/HUMAN_9606_idmapping.dat\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>P31946</th>\n",
       "      <th>UniProtKB-ID</th>\n",
       "      <th>1433B_HUMAN</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P31946</td>\n",
       "      <td>Gene_Name</td>\n",
       "      <td>YWHAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P31946</td>\n",
       "      <td>GI</td>\n",
       "      <td>4507949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>P31946</td>\n",
       "      <td>GI</td>\n",
       "      <td>377656702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>P31946</td>\n",
       "      <td>GI</td>\n",
       "      <td>67464628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P31946</td>\n",
       "      <td>GI</td>\n",
       "      <td>1345590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6288915</th>\n",
       "      <td>A0A0K0L4Y8</td>\n",
       "      <td>EMBL</td>\n",
       "      <td>KF833265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6288916</th>\n",
       "      <td>A0A0K0L4Y8</td>\n",
       "      <td>EMBL-CDS</td>\n",
       "      <td>AIS72444.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6288917</th>\n",
       "      <td>A0A0K0L4Y8</td>\n",
       "      <td>NCBI_TaxID</td>\n",
       "      <td>9606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6288918</th>\n",
       "      <td>A0A0K0L4Y8</td>\n",
       "      <td>ChiTaRS</td>\n",
       "      <td>PNPLA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6288919</th>\n",
       "      <td>A0A0K0L4Y8</td>\n",
       "      <td>CRC64</td>\n",
       "      <td>8A326C624F20B553</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6288920 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             P31946 UniProtKB-ID       1433B_HUMAN\n",
       "0            P31946    Gene_Name             YWHAB\n",
       "1            P31946           GI           4507949\n",
       "2            P31946           GI         377656702\n",
       "3            P31946           GI          67464628\n",
       "4            P31946           GI           1345590\n",
       "...             ...          ...               ...\n",
       "6288915  A0A0K0L4Y8         EMBL          KF833265\n",
       "6288916  A0A0K0L4Y8     EMBL-CDS        AIS72444.1\n",
       "6288917  A0A0K0L4Y8   NCBI_TaxID              9606\n",
       "6288918  A0A0K0L4Y8      ChiTaRS            PNPLA2\n",
       "6288919  A0A0K0L4Y8        CRC64  8A326C624F20B553\n",
       "\n",
       "[6288920 rows x 3 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot_human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save a list of all the symbols in order to search their uniprot\n",
    "sym_to_fix=[]\n",
    "sym_to_fix.extend(biogrid_human['Official Symbol Interactor A'])\n",
    "sym_to_fix.extend(biogrid_human['Official Symbol Interactor B'])\n",
    "\n",
    "#remove duplicates\n",
    "sym_to_fix=list(set(sym_to_fix))\n",
    "\n",
    "# using join() \n",
    "# avoiding printing last comma \n",
    "print(\"The formatted output is : \") \n",
    "print(', '.join(sym_to_fix)) \n",
    "##print in order to search on uniprot.com"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>yourlist:M202001096746803381A1F0E0DB47453E0216320D5745D00</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Entry name</th>\n",
       "      <th>Status</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Annotation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SLC6A1</td>\n",
       "      <td>P30531</td>\n",
       "      <td>SC6A1_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Sodium- and chloride-dependent GABA transporte...</td>\n",
       "      <td>SLC6A1 GABATR GABT1 GAT1</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>5 out of 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRSS48</td>\n",
       "      <td>Q7RTY5</td>\n",
       "      <td>PRS48_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Serine protease 48 (EC 3.4.21.-) (Epidermis-sp...</td>\n",
       "      <td>PRSS48 ESSPL</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>4 out of 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSN1</td>\n",
       "      <td>Q9H410</td>\n",
       "      <td>DSN1_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Kinetochore-associated protein DSN1 homolog</td>\n",
       "      <td>DSN1 C20orf172 MIS13</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>5 out of 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RPL32</td>\n",
       "      <td>P62910</td>\n",
       "      <td>RL32_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>60S ribosomal protein L32 (Large ribosomal sub...</td>\n",
       "      <td>RPL32 PP9932</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>5 out of 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>IFT122</td>\n",
       "      <td>Q9HBG6</td>\n",
       "      <td>IF122_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Intraflagellar transport protein 122 homolog (...</td>\n",
       "      <td>IFT122 SPG WDR10 WDR140</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>5 out of 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16933</th>\n",
       "      <td>MMAA,SLC7A5P2</td>\n",
       "      <td>Q9GIP4</td>\n",
       "      <td>LAT1L_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Putative L-type amino acid transporter 1-like ...</td>\n",
       "      <td>SLC7A5P2 IMAA MMAA</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>3 out of 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16934</th>\n",
       "      <td>SNAPC1</td>\n",
       "      <td>Q16533</td>\n",
       "      <td>SNPC1_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>snRNA-activating protein complex subunit 1 (SN...</td>\n",
       "      <td>SNAPC1 SNAP43</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>4 out of 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16935</th>\n",
       "      <td>DOCK5</td>\n",
       "      <td>Q9H7D0</td>\n",
       "      <td>DOCK5_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Dedicator of cytokinesis protein 5</td>\n",
       "      <td>DOCK5</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>5 out of 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16936</th>\n",
       "      <td>RNF19B</td>\n",
       "      <td>Q6ZMZ0</td>\n",
       "      <td>RN19B_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>E3 ubiquitin-protein ligase RNF19B (EC 2.3.2.3...</td>\n",
       "      <td>RNF19B IBRDC3 NKLAM</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>5 out of 5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16937</th>\n",
       "      <td>SLC6A3</td>\n",
       "      <td>Q01959</td>\n",
       "      <td>SC6A3_HUMAN</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>Sodium-dependent dopamine transporter (DA tran...</td>\n",
       "      <td>SLC6A3 DAT1</td>\n",
       "      <td>Homo sapiens (Human)</td>\n",
       "      <td>5 out of 5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16938 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      yourlist:M202001096746803381A1F0E0DB47453E0216320D5745D00   Entry  \\\n",
       "0                                                 SLC6A1         P30531   \n",
       "1                                                 PRSS48         Q7RTY5   \n",
       "2                                                   DSN1         Q9H410   \n",
       "3                                                  RPL32         P62910   \n",
       "4                                                 IFT122         Q9HBG6   \n",
       "...                                                  ...            ...   \n",
       "16933                                      MMAA,SLC7A5P2         Q9GIP4   \n",
       "16934                                             SNAPC1         Q16533   \n",
       "16935                                              DOCK5         Q9H7D0   \n",
       "16936                                             RNF19B         Q6ZMZ0   \n",
       "16937                                             SLC6A3         Q01959   \n",
       "\n",
       "        Entry name    Status  \\\n",
       "0      SC6A1_HUMAN  reviewed   \n",
       "1      PRS48_HUMAN  reviewed   \n",
       "2       DSN1_HUMAN  reviewed   \n",
       "3       RL32_HUMAN  reviewed   \n",
       "4      IF122_HUMAN  reviewed   \n",
       "...            ...       ...   \n",
       "16933  LAT1L_HUMAN  reviewed   \n",
       "16934  SNPC1_HUMAN  reviewed   \n",
       "16935  DOCK5_HUMAN  reviewed   \n",
       "16936  RN19B_HUMAN  reviewed   \n",
       "16937  SC6A3_HUMAN  reviewed   \n",
       "\n",
       "                                           Protein names  \\\n",
       "0      Sodium- and chloride-dependent GABA transporte...   \n",
       "1      Serine protease 48 (EC 3.4.21.-) (Epidermis-sp...   \n",
       "2            Kinetochore-associated protein DSN1 homolog   \n",
       "3      60S ribosomal protein L32 (Large ribosomal sub...   \n",
       "4      Intraflagellar transport protein 122 homolog (...   \n",
       "...                                                  ...   \n",
       "16933  Putative L-type amino acid transporter 1-like ...   \n",
       "16934  snRNA-activating protein complex subunit 1 (SN...   \n",
       "16935                 Dedicator of cytokinesis protein 5   \n",
       "16936  E3 ubiquitin-protein ligase RNF19B (EC 2.3.2.3...   \n",
       "16937  Sodium-dependent dopamine transporter (DA tran...   \n",
       "\n",
       "                     Gene names              Organism  Annotation  \n",
       "0      SLC6A1 GABATR GABT1 GAT1  Homo sapiens (Human)  5 out of 5  \n",
       "1                  PRSS48 ESSPL  Homo sapiens (Human)  4 out of 5  \n",
       "2          DSN1 C20orf172 MIS13  Homo sapiens (Human)  5 out of 5  \n",
       "3                  RPL32 PP9932  Homo sapiens (Human)  5 out of 5  \n",
       "4       IFT122 SPG WDR10 WDR140  Homo sapiens (Human)  5 out of 5  \n",
       "...                         ...                   ...         ...  \n",
       "16933        SLC7A5P2 IMAA MMAA  Homo sapiens (Human)  3 out of 5  \n",
       "16934             SNAPC1 SNAP43  Homo sapiens (Human)  4 out of 5  \n",
       "16935                     DOCK5  Homo sapiens (Human)  5 out of 5  \n",
       "16936       RNF19B IBRDC3 NKLAM  Homo sapiens (Human)  5 out of 5  \n",
       "16937               SLC6A3 DAT1  Homo sapiens (Human)  5 out of 5  \n",
       "\n",
       "[16938 rows x 8 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#upload the uniprot fixing file\n",
    "unigene = pd.read_csv(\"./uniprot/geneid-uniprot.tab\", sep = '\\t')\n",
    "unigene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#BioGRID Interaction ID</th>\n",
       "      <th>Entrez Gene Interactor A</th>\n",
       "      <th>Entrez Gene Interactor B</th>\n",
       "      <th>BioGRID ID Interactor A</th>\n",
       "      <th>BioGRID ID Interactor B</th>\n",
       "      <th>Systematic Name Interactor A</th>\n",
       "      <th>Systematic Name Interactor B</th>\n",
       "      <th>Official Symbol Interactor A</th>\n",
       "      <th>Official Symbol Interactor B</th>\n",
       "      <th>Synonyms Interactor A</th>\n",
       "      <th>...</th>\n",
       "      <th>Organism Interactor B</th>\n",
       "      <th>Throughput</th>\n",
       "      <th>Score</th>\n",
       "      <th>Modification</th>\n",
       "      <th>Phenotypes</th>\n",
       "      <th>Qualifications</th>\n",
       "      <th>Tags</th>\n",
       "      <th>Source Database</th>\n",
       "      <th>UniprotAC interactor A</th>\n",
       "      <th>UniprotAC interactor B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>103</td>\n",
       "      <td>6416</td>\n",
       "      <td>2318</td>\n",
       "      <td>112315</td>\n",
       "      <td>108607</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MAP2K4</td>\n",
       "      <td>FLNC</td>\n",
       "      <td>JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...</td>\n",
       "      <td>...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P45985</td>\n",
       "      <td>Q14315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>117</td>\n",
       "      <td>84665</td>\n",
       "      <td>88</td>\n",
       "      <td>124185</td>\n",
       "      <td>106603</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MYPN</td>\n",
       "      <td>ACTN2</td>\n",
       "      <td>CMD1DD|CMH22|MYOP|RCM4</td>\n",
       "      <td>...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q86TC9</td>\n",
       "      <td>P35609</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>183</td>\n",
       "      <td>90</td>\n",
       "      <td>2339</td>\n",
       "      <td>106605</td>\n",
       "      <td>108625</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ACVR1</td>\n",
       "      <td>FNTA</td>\n",
       "      <td>ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI</td>\n",
       "      <td>...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q04771</td>\n",
       "      <td>P49354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>278</td>\n",
       "      <td>2624</td>\n",
       "      <td>5371</td>\n",
       "      <td>108894</td>\n",
       "      <td>111384</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>GATA2</td>\n",
       "      <td>PML</td>\n",
       "      <td>DCML|IMD21|MONOMAC|NFE1B</td>\n",
       "      <td>...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P23769</td>\n",
       "      <td>P29590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>418</td>\n",
       "      <td>6118</td>\n",
       "      <td>6774</td>\n",
       "      <td>112038</td>\n",
       "      <td>112651</td>\n",
       "      <td>RP4-547C9.3</td>\n",
       "      <td>-</td>\n",
       "      <td>RPA2</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>REPA2|RP-A p32|RP-A p34|RPA32</td>\n",
       "      <td>...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>P15927</td>\n",
       "      <td>P40763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485375</th>\n",
       "      <td>2621674</td>\n",
       "      <td>84376</td>\n",
       "      <td>10749</td>\n",
       "      <td>124068</td>\n",
       "      <td>115972</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>HOOK3</td>\n",
       "      <td>KIF1C</td>\n",
       "      <td>HK3</td>\n",
       "      <td>...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q86VS8</td>\n",
       "      <td>O43896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485376</th>\n",
       "      <td>2621675</td>\n",
       "      <td>2059</td>\n",
       "      <td>50618</td>\n",
       "      <td>108373</td>\n",
       "      <td>119098</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>EPS8</td>\n",
       "      <td>ITSN2</td>\n",
       "      <td>DFNB102</td>\n",
       "      <td>...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q12929</td>\n",
       "      <td>Q9NZM3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485377</th>\n",
       "      <td>2621676</td>\n",
       "      <td>2059</td>\n",
       "      <td>50618</td>\n",
       "      <td>108373</td>\n",
       "      <td>119098</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>EPS8</td>\n",
       "      <td>ITSN2</td>\n",
       "      <td>DFNB102</td>\n",
       "      <td>...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q12929</td>\n",
       "      <td>Q9NZM3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485378</th>\n",
       "      <td>2621677</td>\n",
       "      <td>50618</td>\n",
       "      <td>2059</td>\n",
       "      <td>119098</td>\n",
       "      <td>108373</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ITSN2</td>\n",
       "      <td>EPS8</td>\n",
       "      <td>PRO2015|SH3D1B|SH3P18|SWA|SWAP</td>\n",
       "      <td>...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q9NZM3</td>\n",
       "      <td>Q12929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485379</th>\n",
       "      <td>2621678</td>\n",
       "      <td>2059</td>\n",
       "      <td>50618</td>\n",
       "      <td>108373</td>\n",
       "      <td>119098</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>EPS8</td>\n",
       "      <td>ITSN2</td>\n",
       "      <td>DFNB102</td>\n",
       "      <td>...</td>\n",
       "      <td>9606</td>\n",
       "      <td>Low Throughput</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>BIOGRID</td>\n",
       "      <td>Q12929</td>\n",
       "      <td>Q9NZM3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>485380 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        #BioGRID Interaction ID Entrez Gene Interactor A  \\\n",
       "0                           103                     6416   \n",
       "1                           117                    84665   \n",
       "2                           183                       90   \n",
       "3                           278                     2624   \n",
       "4                           418                     6118   \n",
       "...                         ...                      ...   \n",
       "485375                  2621674                    84376   \n",
       "485376                  2621675                     2059   \n",
       "485377                  2621676                     2059   \n",
       "485378                  2621677                    50618   \n",
       "485379                  2621678                     2059   \n",
       "\n",
       "       Entrez Gene Interactor B  BioGRID ID Interactor A  \\\n",
       "0                          2318                   112315   \n",
       "1                            88                   124185   \n",
       "2                          2339                   106605   \n",
       "3                          5371                   108894   \n",
       "4                          6774                   112038   \n",
       "...                         ...                      ...   \n",
       "485375                    10749                   124068   \n",
       "485376                    50618                   108373   \n",
       "485377                    50618                   108373   \n",
       "485378                     2059                   119098   \n",
       "485379                    50618                   108373   \n",
       "\n",
       "        BioGRID ID Interactor B Systematic Name Interactor A  \\\n",
       "0                        108607                            -   \n",
       "1                        106603                            -   \n",
       "2                        108625                            -   \n",
       "3                        111384                            -   \n",
       "4                        112651                  RP4-547C9.3   \n",
       "...                         ...                          ...   \n",
       "485375                   115972                            -   \n",
       "485376                   119098                            -   \n",
       "485377                   119098                            -   \n",
       "485378                   108373                            -   \n",
       "485379                   119098                            -   \n",
       "\n",
       "       Systematic Name Interactor B Official Symbol Interactor A  \\\n",
       "0                                 -                       MAP2K4   \n",
       "1                                 -                         MYPN   \n",
       "2                                 -                        ACVR1   \n",
       "3                                 -                        GATA2   \n",
       "4                                 -                         RPA2   \n",
       "...                             ...                          ...   \n",
       "485375                            -                        HOOK3   \n",
       "485376                            -                         EPS8   \n",
       "485377                            -                         EPS8   \n",
       "485378                            -                        ITSN2   \n",
       "485379                            -                         EPS8   \n",
       "\n",
       "       Official Symbol Interactor B  \\\n",
       "0                              FLNC   \n",
       "1                             ACTN2   \n",
       "2                              FNTA   \n",
       "3                               PML   \n",
       "4                             STAT3   \n",
       "...                             ...   \n",
       "485375                        KIF1C   \n",
       "485376                        ITSN2   \n",
       "485377                        ITSN2   \n",
       "485378                         EPS8   \n",
       "485379                        ITSN2   \n",
       "\n",
       "                                    Synonyms Interactor A  ...  \\\n",
       "0       JNKK|JNKK1|MAPKK4|MEK4|MKK4|PRKMK4|SAPKK-1|SAP...  ...   \n",
       "1                                  CMD1DD|CMH22|MYOP|RCM4  ...   \n",
       "2                 ACTRI|ACVR1A|ACVRLK2|ALK2|FOP|SKR1|TSRI  ...   \n",
       "3                                DCML|IMD21|MONOMAC|NFE1B  ...   \n",
       "4                           REPA2|RP-A p32|RP-A p34|RPA32  ...   \n",
       "...                                                   ...  ...   \n",
       "485375                                                HK3  ...   \n",
       "485376                                            DFNB102  ...   \n",
       "485377                                            DFNB102  ...   \n",
       "485378                     PRO2015|SH3D1B|SH3P18|SWA|SWAP  ...   \n",
       "485379                                            DFNB102  ...   \n",
       "\n",
       "       Organism Interactor B      Throughput Score Modification  Phenotypes  \\\n",
       "0                       9606  Low Throughput     -            -           -   \n",
       "1                       9606  Low Throughput     -            -           -   \n",
       "2                       9606  Low Throughput     -            -           -   \n",
       "3                       9606  Low Throughput     -            -           -   \n",
       "4                       9606  Low Throughput     -            -           -   \n",
       "...                      ...             ...   ...          ...         ...   \n",
       "485375                  9606  Low Throughput     -            -           -   \n",
       "485376                  9606  Low Throughput     -            -           -   \n",
       "485377                  9606  Low Throughput     -            -           -   \n",
       "485378                  9606  Low Throughput     -            -           -   \n",
       "485379                  9606  Low Throughput     -            -           -   \n",
       "\n",
       "        Qualifications  Tags Source Database UniprotAC interactor A  \\\n",
       "0                    -     -         BIOGRID                 P45985   \n",
       "1                    -     -         BIOGRID                 Q86TC9   \n",
       "2                    -     -         BIOGRID                 Q04771   \n",
       "3                    -     -         BIOGRID                 P23769   \n",
       "4                    -     -         BIOGRID                 P15927   \n",
       "...                ...   ...             ...                    ...   \n",
       "485375               -     -         BIOGRID                 Q86VS8   \n",
       "485376               -     -         BIOGRID                 Q12929   \n",
       "485377               -     -         BIOGRID                 Q12929   \n",
       "485378               -     -         BIOGRID                 Q9NZM3   \n",
       "485379               -     -         BIOGRID                 Q12929   \n",
       "\n",
       "       UniprotAC interactor B  \n",
       "0                      Q14315  \n",
       "1                      P35609  \n",
       "2                      P49354  \n",
       "3                      P29590  \n",
       "4                      P40763  \n",
       "...                       ...  \n",
       "485375                 O43896  \n",
       "485376                 Q9NZM3  \n",
       "485377                 Q9NZM3  \n",
       "485378                 Q12929  \n",
       "485379                 Q9NZM3  \n",
       "\n",
       "[485380 rows x 26 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#create a dictionary that maps symbol with its uniprot\n",
    "unigene=unigene.rename(columns={\"yourlist:M202001096746803381A1F0E0DB47453E0216320D5745D00\": \"symbol\"})\n",
    "unigene=pd.Series(unigene.Entry.values, index=unigene.symbol).to_dict()\n",
    "biogrid_human['UniprotAC interactor A']= biogrid_human['Official Symbol Interactor A'].map(unigene)\n",
    "biogrid_human['UniprotAC interactor B']= biogrid_human['Official Symbol Interactor B'].map(unigene)\n",
    "biogrid_human"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot1</th>\n",
       "      <th>uniprot2</th>\n",
       "      <th>symbol1</th>\n",
       "      <th>symbol2</th>\n",
       "      <th>evidence type</th>\n",
       "      <th>cancer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Q9NUX5</td>\n",
       "      <td>Q9NVM4</td>\n",
       "      <td>POT1</td>\n",
       "      <td>PRMT7</td>\n",
       "      <td>exp</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P62633</td>\n",
       "      <td>Q99729</td>\n",
       "      <td>CNBP</td>\n",
       "      <td>HNRNPAB</td>\n",
       "      <td>exp</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>O43715</td>\n",
       "      <td>O95817</td>\n",
       "      <td>TRIAP1</td>\n",
       "      <td>BAG3</td>\n",
       "      <td>exp</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q13285</td>\n",
       "      <td>Q9UKM9</td>\n",
       "      <td>NR5A1</td>\n",
       "      <td>RALY</td>\n",
       "      <td>exp</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>P60174</td>\n",
       "      <td>Q9H0W5</td>\n",
       "      <td>TPI1</td>\n",
       "      <td>CCDC8</td>\n",
       "      <td>exp</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272487</th>\n",
       "      <td>O60341</td>\n",
       "      <td>Q8NEZ2</td>\n",
       "      <td>KDM1A</td>\n",
       "      <td>VPS37A</td>\n",
       "      <td>exp</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272488</th>\n",
       "      <td>Q00987</td>\n",
       "      <td>Q8TAQ2</td>\n",
       "      <td>MDM2</td>\n",
       "      <td>SMARCC2</td>\n",
       "      <td>exp</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272489</th>\n",
       "      <td>P22314</td>\n",
       "      <td>Q5VTR2</td>\n",
       "      <td>UBA1</td>\n",
       "      <td>RNF20</td>\n",
       "      <td>exp</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272490</th>\n",
       "      <td>Q86X19</td>\n",
       "      <td>Q96J84</td>\n",
       "      <td>TMEM17</td>\n",
       "      <td>KIRREL1</td>\n",
       "      <td>exp</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272491</th>\n",
       "      <td>P40227</td>\n",
       "      <td>Q7Z6Z7</td>\n",
       "      <td>CCT6A</td>\n",
       "      <td>HUWE1</td>\n",
       "      <td>exp</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>272492 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       uniprot1 uniprot2 symbol1  symbol2 evidence type  cancer\n",
       "0        Q9NUX5   Q9NVM4    POT1    PRMT7           exp       0\n",
       "1        P62633   Q99729    CNBP  HNRNPAB           exp       0\n",
       "2        O43715   O95817  TRIAP1     BAG3           exp       0\n",
       "3        Q13285   Q9UKM9   NR5A1     RALY           exp       0\n",
       "4        P60174   Q9H0W5    TPI1    CCDC8           exp       0\n",
       "...         ...      ...     ...      ...           ...     ...\n",
       "272487   O60341   Q8NEZ2   KDM1A   VPS37A           exp       0\n",
       "272488   Q00987   Q8TAQ2    MDM2  SMARCC2           exp       0\n",
       "272489   P22314   Q5VTR2    UBA1    RNF20           exp       0\n",
       "272490   Q86X19   Q96J84  TMEM17  KIRREL1           exp       0\n",
       "272491   P40227   Q7Z6Z7   CCT6A    HUWE1           exp       0\n",
       "\n",
       "[272492 rows x 6 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "iid_human = pd.read_csv('iid.txt', sep = '\\t')\n",
    "iid_human.drop(['Unnamed: 0'], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SEED GENES INTERACTOME\n",
    "def build_first_table(biogrid_human, iid_human):\n",
    "    db1 = 'Biogrid Human'\n",
    "    db2 = 'Integrated Interactions Database experimental data'\n",
    "    t = pd.DataFrame(columns=['interactorA', 'interactorB', \n",
    "                                        'interactorA_Uniprot_AC', 'interactorB_Uniprot_AC', 'db_source'])\n",
    "    for i in range(len(biogrid_human)):\n",
    "        sa = biogrid_human['Official Symbol Interactor A'][i]\n",
    "        sb = biogrid_human['Official Symbol Interactor B'][i]\n",
    "        uniprota = biogrid_human['UniprotAC interactor A'][i]\n",
    "        uniprotb = biogrid_human['UniprotAC interactor B'][i]\n",
    "        if sa in geneSymbol and sb in geneSymbol:\n",
    "            t = t.append({'interactorA':sa, 'interactorB':sb, \n",
    "                          'interactorA_Uniprot_AC':uniprota, 'interactorB_Uniprot_AC':uniprotb, 'db_source': db1}\n",
    "                         , ignore_index=True)\n",
    "    for i in range(len(iid_human)):\n",
    "        sa = iid_human['symbol1'][i]\n",
    "        sb = iid_human['symbol2'][i]\n",
    "        uniprota = iid_human['uniprot1'][i]\n",
    "        uniprotb = iid_human['uniprot2'][i]\n",
    "        if sa in geneSymbol and sb in geneSymbol:\n",
    "            t = t.append({'interactorA':sa, 'interactorB':sb, \n",
    "                          'interactorA_Uniprot_AC':uniprota, 'interactorB_Uniprot_AC':uniprotb, 'db_source': db2}\n",
    "                         , ignore_index=True)\n",
    "    t.to_csv(\"seed_genes_interactome.tsv\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "biogrid_human=biogrid_human.reset_index(drop=True)\n",
    "iid_human=iid_human.reset_index(drop=True)\n",
    "build_first_table(biogrid_human, iid_human)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interactorA</th>\n",
       "      <th>interactorB</th>\n",
       "      <th>interactorA_Uniprot_AC</th>\n",
       "      <th>interactorB_Uniprot_AC</th>\n",
       "      <th>db_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ILK</td>\n",
       "      <td>DDX3X</td>\n",
       "      <td>Q13418</td>\n",
       "      <td>O00571</td>\n",
       "      <td>Biogrid Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ILK</td>\n",
       "      <td>COPG1</td>\n",
       "      <td>Q13418</td>\n",
       "      <td>Q9Y678</td>\n",
       "      <td>Biogrid Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>CDK6</td>\n",
       "      <td>Q8N726</td>\n",
       "      <td>Q00534</td>\n",
       "      <td>Biogrid Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RXRA</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>P19793</td>\n",
       "      <td>P35222</td>\n",
       "      <td>Biogrid Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MUC1</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>P15941</td>\n",
       "      <td>P35222</td>\n",
       "      <td>Biogrid Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>TXN</td>\n",
       "      <td>CDKN1A</td>\n",
       "      <td>P10599</td>\n",
       "      <td>P38936</td>\n",
       "      <td>Integrated Interactions Database experimental ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>TRAF7</td>\n",
       "      <td>TRAF7</td>\n",
       "      <td>Q6Q0C0</td>\n",
       "      <td>Q6Q0C0</td>\n",
       "      <td>Integrated Interactions Database experimental ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>RXRA</td>\n",
       "      <td>IL12B</td>\n",
       "      <td>P19793</td>\n",
       "      <td>P29460</td>\n",
       "      <td>Integrated Interactions Database experimental ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>C9</td>\n",
       "      <td>FN1</td>\n",
       "      <td>P02748</td>\n",
       "      <td>P02751</td>\n",
       "      <td>Integrated Interactions Database experimental ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>NAT2</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>P11245</td>\n",
       "      <td>P42771</td>\n",
       "      <td>Integrated Interactions Database experimental ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>480 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    interactorA interactorB interactorA_Uniprot_AC interactorB_Uniprot_AC  \\\n",
       "0           ILK       DDX3X                 Q13418                 O00571   \n",
       "1           ILK       COPG1                 Q13418                 Q9Y678   \n",
       "2        CDKN2A        CDK6                 Q8N726                 Q00534   \n",
       "3          RXRA      CTNNB1                 P19793                 P35222   \n",
       "4          MUC1      CTNNB1                 P15941                 P35222   \n",
       "..          ...         ...                    ...                    ...   \n",
       "475         TXN      CDKN1A                 P10599                 P38936   \n",
       "476       TRAF7       TRAF7                 Q6Q0C0                 Q6Q0C0   \n",
       "477        RXRA       IL12B                 P19793                 P29460   \n",
       "478          C9         FN1                 P02748                 P02751   \n",
       "479        NAT2      CDKN2A                 P11245                 P42771   \n",
       "\n",
       "                                             db_source  \n",
       "0                                        Biogrid Human  \n",
       "1                                        Biogrid Human  \n",
       "2                                        Biogrid Human  \n",
       "3                                        Biogrid Human  \n",
       "4                                        Biogrid Human  \n",
       "..                                                 ...  \n",
       "475  Integrated Interactions Database experimental ...  \n",
       "476  Integrated Interactions Database experimental ...  \n",
       "477  Integrated Interactions Database experimental ...  \n",
       "478  Integrated Interactions Database experimental ...  \n",
       "479  Integrated Interactions Database experimental ...  \n",
       "\n",
       "[480 rows x 5 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interactome_seed = pd.read_csv(\"seed_genes_interactome.tsv\", sep = '\\t')\n",
    "interactome_seed.drop(['Unnamed: 0'], axis = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "union interactome: all proteins interacting with at least one seed gene, from all DBs, same format as above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "biogrid_human.to_csv(\"new_biogrid.tsv\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_union_interactome(biogrid_human, iid_human):\n",
    "    db1 = 'Biogrid Human'\n",
    "    db2 = 'Integrated Interactions Database experimental data'\n",
    "    t = pd.DataFrame(columns=['interactorA', 'interactorB', \n",
    "                                        'interactorA_Uniprot_AC', 'interactorB_Uniprot_AC', 'db_source'])\n",
    "    for i in range(len(biogrid_human)):\n",
    "        sa = biogrid_human['Official Symbol Interactor A'][i]\n",
    "        sb = biogrid_human['Official Symbol Interactor B'][i]\n",
    "        uniprota = biogrid_human['UniprotAC interactor A'][i]\n",
    "        uniprotb = biogrid_human['UniprotAC interactor B'][i]\n",
    "        if sa in geneSymbol or sb in geneSymbol:\n",
    "            t = t.append({'interactorA':sa, 'interactorB':sb, \n",
    "                          'interactorA_Uniprot_AC':uniprota, 'interactorB_Uniprot_AC':uniprotb, 'db_source': db1}\n",
    "                         , ignore_index=True)\n",
    "    for i in range(len(iid_human)):\n",
    "        sa = iid_human['symbol1'][i]\n",
    "        sb = iid_human['symbol2'][i]\n",
    "        uniprota = iid_human['uniprot1'][i]\n",
    "        uniprotb = iid_human['uniprot2'][i]\n",
    "        if sa in geneSymbol or sb in geneSymbol:\n",
    "            t = t.append({'interactorA':sa, 'interactorB':sb, \n",
    "                          'interactorA_Uniprot_AC':uniprota, 'interactorB_Uniprot_AC':uniprotb, 'db_source': db2}\n",
    "                         , ignore_index=True)\n",
    "    t.to_csv(\"union_interactome.tsv\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "# consider also non/seed-non/seed interactions\n",
    "def build_nonseed_dataframes(biogrid_human, iid_human, union_human):\n",
    "    non_seed_df_A = union_human.loc[~(union_human['interactorA'].isin(geneSymbol))]\n",
    "    non_seed_df_B = union_human.loc[~(union_human['interactorB'].isin(geneSymbol))]\n",
    "    non_seed_df_A=non_seed_df_A.reset_index(drop=True)\n",
    "    non_seed_df_B=non_seed_df_B.reset_index(drop=True)\n",
    "    \n",
    "    #build a list with only non seed genes that interacts with at least one seed gene\n",
    "    non_seed_list=[]\n",
    "    for i in range(0, len(non_seed_df_A)):\n",
    "        non_seed_list.append(non_seed_df_A['interactorA'][i])\n",
    "    for i in range(0, len(non_seed_df_B)):\n",
    "        non_seed_list.append(non_seed_df_B['interactorB'][i])\n",
    "    #drop duplicates\n",
    "    non_seed_list = list(dict.fromkeys(non_seed_list))\n",
    "    list_of_idx=[]\n",
    "    #fill the list\n",
    "    for i in range(len(biogrid_human)):\n",
    "        if biogrid_human['Official Symbol Interactor A'][i] in non_seed_list and biogrid_human['Official Symbol Interactor B'][i] in non_seed_list:\n",
    "            list_of_idx.append(i)\n",
    "\n",
    "    biogrid_non_seed = biogrid_human.loc[list_of_idx]\n",
    "    biogrid_non_seed = biogrid_non_seed.reset_index(drop=True)\n",
    "    biogrid_non_seed.to_csv(\"biogrid_union_non_seed.tsv\", sep = '\\t')\n",
    "\n",
    "    #same procedure for iid\n",
    "    list_of_idx=[]\n",
    "    #fill the list\n",
    "    for i in range(len(iid_human)):\n",
    "        if iid_human['symbol1'][i] in non_seed_list and iid_human['symbol2'][i] in non_seed_list:\n",
    "            list_of_idx.append(i)\n",
    "\n",
    "    iid_non_seed = iid_human.loc[list_of_idx]\n",
    "    iid_non_seed = iid_non_seed.reset_index(drop=True)                  \n",
    "    iid_non_seed.to_csv(\"iid_union_non_seed.tsv\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_non_seed_dataframes(biogrid_human, iid_human, union_human):\n",
    "    db1 = 'Biogrid Human'\n",
    "    db2 = 'Integrated Interactions Database experimental data'\n",
    "    union_dict = {}\n",
    "    \n",
    "    for i in range(len(iid_human)):\n",
    "        sa = iid_human['symbol1'][i]\n",
    "        sb = iid_human['symbol2'][i]\n",
    "        uniprota = iid_human['uniprot1'][i]\n",
    "        uniprotb = iid_human['uniprot2'][i]\n",
    "        '''union_human = union_human.append({'interactorA':sa, 'interactorB':sb, \n",
    "                          'interactorA_Uniprot_AC':uniprota, 'interactorB_Uniprot_AC':uniprotb, 'db_source': db2}\n",
    "                         , ignore_index=True)'''\n",
    "        \n",
    "        union_dict[i] = {\"interactorA\": sa, 'interactorB':sb, \n",
    "                          'interactorA_Uniprot_AC':uniprota, 'interactorB_Uniprot_AC':uniprotb, 'db_source': db2}\n",
    "\n",
    "    df = pd.DataFrame.from_dict(union_dict, \"index\")  \n",
    "    new = pd.concat([union_human, df], ignore_index=True)   \n",
    "        \n",
    "    print(\"done with iid\", len(union_human))\n",
    "    \n",
    "    union_dict = {} \n",
    "    for i in range(len(biogrid_human)):\n",
    "        sa = biogrid_human['Official Symbol Interactor A'][i]\n",
    "        sb = biogrid_human['Official Symbol Interactor B'][i]\n",
    "        uniprota = biogrid_human['UniprotAC interactor A'][i]\n",
    "        uniprotb = biogrid_human['UniprotAC interactor B'][i]\n",
    "        '''union_human = union_human.append({'interactorA':sa, 'interactorB':sb, \n",
    "                          'interactorA_Uniprot_AC':uniprota, 'interactorB_Uniprot_AC':uniprotb, 'db_source': db1}\n",
    "                         , ignore_index=True)'''\n",
    "        \n",
    "        union_dict[i] = {\"interactorA\": sa, 'interactorB':sb, \n",
    "                          'interactorA_Uniprot_AC':uniprota, 'interactorB_Uniprot_AC':uniprotb, 'db_source': db1}\n",
    "\n",
    "    df = pd.DataFrame.from_dict(union_dict, \"index\")  \n",
    "    new = pd.concat([new, df], ignore_index=True)\n",
    "\n",
    "    # important to set the 'orient' parameter to \"index\" to make the keys as rows\n",
    "    \n",
    "\n",
    "    # drop duplicates for interactorA and interactorB \n",
    "    new = new.drop_duplicates(subset = ['interactorA', 'interactorB'], keep='first')\n",
    "    new.to_csv(\"union_interactome_extended.tsv\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "biogrid_human=biogrid_human.reset_index(drop=True)\n",
    "iid_human=iid_human.reset_index(drop=True)\n",
    "#build_union_interactome(biogrid_human, iid_human)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/giulia/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:19: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done with iid 25039\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/giulia/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:37: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version\n",
      "of pandas will change to not sort by default.\n",
      "\n",
      "To accept the future behavior, pass 'sort=False'.\n",
      "\n",
      "To retain the current behavior and silence the warning, pass 'sort=True'.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "union_human = pd.read_csv(\"union_interactome.tsv\", sep = '\\t')\n",
    "biogrid_human=biogrid_human.reset_index(drop=True)\n",
    "iid_human=iid_human.reset_index(drop=True)\n",
    "union_human=union_human.reset_index(drop=True)\n",
    "build_nonseed_dataframes(biogrid_human, iid_human, union_human)\n",
    "\n",
    "nsbio = pd.read_csv(\"biogrid_union_non_seed.tsv\", sep = '\\t')\n",
    "nsiid = pd.read_csv(\"iid_union_non_seed.tsv\", sep = '\\t')\n",
    "nsiid = nsiid.reset_index(drop=True)\n",
    "nsbio = nsbio.reset_index(drop=True)\n",
    "\n",
    "merge_non_seed_dataframes(nsbio, nsiid, union_human)\n",
    "#unionext = pd.read_csv(\"union_interactome_extended.tsv\", sep = '\\t')\n",
    "#union_interactome_extended(biogrid_human, iid_human, union_human)\n",
    "#union_extended = pd.read_csv(\"union_extended.tsv\", sep = '\\t') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>db_source</th>\n",
       "      <th>interactorA</th>\n",
       "      <th>interactorA_Uniprot_AC</th>\n",
       "      <th>interactorB</th>\n",
       "      <th>interactorB_Uniprot_AC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Biogrid Human</td>\n",
       "      <td>MAGI1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>P35222</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Biogrid Human</td>\n",
       "      <td>TSG101</td>\n",
       "      <td>Q99816</td>\n",
       "      <td>CDKN1A</td>\n",
       "      <td>P38936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Biogrid Human</td>\n",
       "      <td>TXN</td>\n",
       "      <td>P10599</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>P02452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Biogrid Human</td>\n",
       "      <td>DCN</td>\n",
       "      <td>P07585</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>P00533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Biogrid Human</td>\n",
       "      <td>HTT</td>\n",
       "      <td>P42858</td>\n",
       "      <td>SETD2</td>\n",
       "      <td>Q9BYW2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267148</th>\n",
       "      <td>Biogrid Human</td>\n",
       "      <td>NINL</td>\n",
       "      <td>Q9Y2I6</td>\n",
       "      <td>MAML1</td>\n",
       "      <td>Q92585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267149</th>\n",
       "      <td>Biogrid Human</td>\n",
       "      <td>NINL</td>\n",
       "      <td>Q9Y2I6</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>P51587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267150</th>\n",
       "      <td>Biogrid Human</td>\n",
       "      <td>NINL</td>\n",
       "      <td>Q9Y2I6</td>\n",
       "      <td>NKAP</td>\n",
       "      <td>Q8N5F7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267151</th>\n",
       "      <td>Biogrid Human</td>\n",
       "      <td>EPS8</td>\n",
       "      <td>Q12929</td>\n",
       "      <td>ITSN2</td>\n",
       "      <td>Q9NZM3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>267152</th>\n",
       "      <td>Biogrid Human</td>\n",
       "      <td>ITSN2</td>\n",
       "      <td>Q9NZM3</td>\n",
       "      <td>EPS8</td>\n",
       "      <td>Q12929</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>267153 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            db_source interactorA interactorA_Uniprot_AC interactorB  \\\n",
       "0       Biogrid Human       MAGI1                    NaN      CTNNB1   \n",
       "1       Biogrid Human      TSG101                 Q99816      CDKN1A   \n",
       "2       Biogrid Human         TXN                 P10599      COL1A1   \n",
       "3       Biogrid Human         DCN                 P07585        EGFR   \n",
       "4       Biogrid Human         HTT                 P42858       SETD2   \n",
       "...               ...         ...                    ...         ...   \n",
       "267148  Biogrid Human        NINL                 Q9Y2I6       MAML1   \n",
       "267149  Biogrid Human        NINL                 Q9Y2I6       BRCA2   \n",
       "267150  Biogrid Human        NINL                 Q9Y2I6        NKAP   \n",
       "267151  Biogrid Human        EPS8                 Q12929       ITSN2   \n",
       "267152  Biogrid Human       ITSN2                 Q9NZM3        EPS8   \n",
       "\n",
       "       interactorB_Uniprot_AC  \n",
       "0                      P35222  \n",
       "1                      P38936  \n",
       "2                      P02452  \n",
       "3                      P00533  \n",
       "4                      Q9BYW2  \n",
       "...                       ...  \n",
       "267148                 Q92585  \n",
       "267149                 P51587  \n",
       "267150                 Q8N5F7  \n",
       "267151                 Q9NZM3  \n",
       "267152                 Q12929  \n",
       "\n",
       "[267153 rows x 5 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#drop duplicates\n",
    "new = pd.read_csv(\"union_interactome_extended.tsv\", sep = '\\t')\n",
    "new = new.drop(['Unnamed: 0', 'Unnamed: 0.1'], axis = 1)\n",
    "new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "new.to_csv(\"union_interactome_extended.tsv\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>interactorA</th>\n",
       "      <th>interactorB</th>\n",
       "      <th>interactorA_Uniprot_AC</th>\n",
       "      <th>interactorB_Uniprot_AC</th>\n",
       "      <th>db_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>MAGI1</td>\n",
       "      <td>CTNNB1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P35222</td>\n",
       "      <td>Biogrid Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TSG101</td>\n",
       "      <td>CDKN1A</td>\n",
       "      <td>Q99816</td>\n",
       "      <td>P38936</td>\n",
       "      <td>Biogrid Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TXN</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>P10599</td>\n",
       "      <td>P02452</td>\n",
       "      <td>Biogrid Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>DCN</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>P07585</td>\n",
       "      <td>P00533</td>\n",
       "      <td>Biogrid Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>HTT</td>\n",
       "      <td>SETD2</td>\n",
       "      <td>P42858</td>\n",
       "      <td>Q9BYW2</td>\n",
       "      <td>Biogrid Human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25034</th>\n",
       "      <td>25034</td>\n",
       "      <td>LTA</td>\n",
       "      <td>ANKRD40</td>\n",
       "      <td>P01374</td>\n",
       "      <td>Q6AI12</td>\n",
       "      <td>Integrated Interactions Database experimental ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25035</th>\n",
       "      <td>25035</td>\n",
       "      <td>FN1</td>\n",
       "      <td>TARDBP</td>\n",
       "      <td>P02751</td>\n",
       "      <td>Q13148</td>\n",
       "      <td>Integrated Interactions Database experimental ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25036</th>\n",
       "      <td>25036</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>PSMD2</td>\n",
       "      <td>P00533</td>\n",
       "      <td>Q13200</td>\n",
       "      <td>Integrated Interactions Database experimental ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25037</th>\n",
       "      <td>25037</td>\n",
       "      <td>TNF</td>\n",
       "      <td>AGO2</td>\n",
       "      <td>P01375</td>\n",
       "      <td>Q9UKV8</td>\n",
       "      <td>Integrated Interactions Database experimental ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25038</th>\n",
       "      <td>25038</td>\n",
       "      <td>DDX3X</td>\n",
       "      <td>CENPA</td>\n",
       "      <td>O00571</td>\n",
       "      <td>P49450</td>\n",
       "      <td>Integrated Interactions Database experimental ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25039 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0 interactorA interactorB interactorA_Uniprot_AC  \\\n",
       "0               0       MAGI1      CTNNB1                    NaN   \n",
       "1               1      TSG101      CDKN1A                 Q99816   \n",
       "2               2         TXN      COL1A1                 P10599   \n",
       "3               3         DCN        EGFR                 P07585   \n",
       "4               4         HTT       SETD2                 P42858   \n",
       "...           ...         ...         ...                    ...   \n",
       "25034       25034         LTA     ANKRD40                 P01374   \n",
       "25035       25035         FN1      TARDBP                 P02751   \n",
       "25036       25036        EGFR       PSMD2                 P00533   \n",
       "25037       25037         TNF        AGO2                 P01375   \n",
       "25038       25038       DDX3X       CENPA                 O00571   \n",
       "\n",
       "      interactorB_Uniprot_AC  \\\n",
       "0                     P35222   \n",
       "1                     P38936   \n",
       "2                     P02452   \n",
       "3                     P00533   \n",
       "4                     Q9BYW2   \n",
       "...                      ...   \n",
       "25034                 Q6AI12   \n",
       "25035                 Q13148   \n",
       "25036                 Q13200   \n",
       "25037                 Q9UKV8   \n",
       "25038                 P49450   \n",
       "\n",
       "                                               db_source  \n",
       "0                                          Biogrid Human  \n",
       "1                                          Biogrid Human  \n",
       "2                                          Biogrid Human  \n",
       "3                                          Biogrid Human  \n",
       "4                                          Biogrid Human  \n",
       "...                                                  ...  \n",
       "25034  Integrated Interactions Database experimental ...  \n",
       "25035  Integrated Interactions Database experimental ...  \n",
       "25036  Integrated Interactions Database experimental ...  \n",
       "25037  Integrated Interactions Database experimental ...  \n",
       "25038  Integrated Interactions Database experimental ...  \n",
       "\n",
       "[25039 rows x 6 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unionint = pd.read_csv(\"union_interactome.tsv\", sep = '\\t')\n",
    "unionint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### intersection interactome: all proteins interacting with at least one seed gene confirmed by both DBs, in the       format: interactor A gene symbol, interactor B gene symbol, interactor A Uniprot AC, interactor B Uniprot AC\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_intersection_interactome(biogrid_human, iid_human):\n",
    "    db1 = 'Biogrid Human'\n",
    "    db2 = 'Integrated Interactions Database experimental data'\n",
    "    union = pd.read_csv(\"union_interactome.tsv\", sep = '\\t')\n",
    "    \n",
    "    union_biogrid = union.loc[(union['db_source'] == 'Biogrid Human')]\n",
    "    union_biogrid = union_biogrid.drop(['Unnamed: 0', 'db_source'], axis = 1)\n",
    "    \n",
    "    union_iid = union.loc[(union['db_source'] == 'Integrated Interactions Database experimental data')]\n",
    "    union_iid = union_iid.drop(['Unnamed: 0', 'db_source'], axis = 1)\n",
    "    \n",
    "    intersect = pd.merge(union_biogrid, union_iid)\n",
    "    intersect.dropna(inplace=True)\n",
    "    intersect.to_csv(\"intersection_interactome.tsv\", sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "biogrid_human=biogrid_human.reset_index(drop=True)\n",
    "iid_human=iid_human.reset_index(drop=True)\n",
    "build_intersection_interactome(biogrid_human, iid_human)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersect = pd.read_csv(\"intersection_interactome.tsv\", sep = '\\t')\n",
    "intersect.drop(['Unnamed: 0'], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>interactorA</th>\n",
       "      <th>interactorB</th>\n",
       "      <th>interactorA_Uniprot_AC</th>\n",
       "      <th>interactorB_Uniprot_AC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>APOA1</td>\n",
       "      <td>PDE1A</td>\n",
       "      <td>P02647</td>\n",
       "      <td>P54750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TXN</td>\n",
       "      <td>NCF4</td>\n",
       "      <td>P10599</td>\n",
       "      <td>Q15080</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>VIM</td>\n",
       "      <td>TCHP</td>\n",
       "      <td>P08670</td>\n",
       "      <td>Q9BT92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABLIM1</td>\n",
       "      <td>VIM</td>\n",
       "      <td>O14639</td>\n",
       "      <td>P08670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>VIM</td>\n",
       "      <td>NIF3L1</td>\n",
       "      <td>P08670</td>\n",
       "      <td>Q9GZT8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3734</th>\n",
       "      <td>TP53</td>\n",
       "      <td>EGLN3</td>\n",
       "      <td>P04637</td>\n",
       "      <td>Q9H6Z9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3735</th>\n",
       "      <td>KIT</td>\n",
       "      <td>PLCE1</td>\n",
       "      <td>P10721</td>\n",
       "      <td>Q9P212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3736</th>\n",
       "      <td>TP53</td>\n",
       "      <td>DNAJB1</td>\n",
       "      <td>P04637</td>\n",
       "      <td>P25685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3737</th>\n",
       "      <td>PHB</td>\n",
       "      <td>AGO2</td>\n",
       "      <td>P35232</td>\n",
       "      <td>Q9UKV8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3738</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>CKAP4</td>\n",
       "      <td>P00533</td>\n",
       "      <td>Q07065</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3739 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     interactorA interactorB interactorA_Uniprot_AC interactorB_Uniprot_AC\n",
       "0          APOA1       PDE1A                 P02647                 P54750\n",
       "1            TXN        NCF4                 P10599                 Q15080\n",
       "2            VIM        TCHP                 P08670                 Q9BT92\n",
       "3         ABLIM1         VIM                 O14639                 P08670\n",
       "4            VIM      NIF3L1                 P08670                 Q9GZT8\n",
       "...          ...         ...                    ...                    ...\n",
       "3734        TP53       EGLN3                 P04637                 Q9H6Z9\n",
       "3735         KIT       PLCE1                 P10721                 Q9P212\n",
       "3736        TP53      DNAJB1                 P04637                 P25685\n",
       "3737         PHB        AGO2                 P35232                 Q9UKV8\n",
       "3738        EGFR       CKAP4                 P00533                 Q07065\n",
       "\n",
       "[3739 rows x 4 columns]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Enrichment analysis\n",
    "* Using the service Enrichr, find, report in tables and save related charts (8 charts in total) of the overrepresented GO categories (limit to the first 10 for each main category, BP, MF, CL) and the the overrepresented pathways (KEGG 2019 Human) for:\n",
    "    * the seed genes,\n",
    "    * the union interactome genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unionint = pd.read_csv(\"union_interactome.tsv\", sep = '\\t')\n",
    "unionint = unionint.drop(['Unnamed: 0'], axis = 1)\n",
    "\n",
    "interactome_seed = pd.read_csv(\"seed_genes_interactome.tsv\", sep = '\\t')\n",
    "interactome_seed = interactome_seed.drop(['Unnamed: 0'], axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "interactome_seed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save a list of all the symbols in order to search their uniprot\n",
    "def print_list(dataframe, column1, column2):\n",
    "    sym_to_fix=[]\n",
    "    sym_to_fix.extend(dataframe[column1])\n",
    "    sym_to_fix.extend(dataframe[column2])\n",
    "    #remove duplicates\n",
    "    sym_to_fix=list(set(sym_to_fix))\n",
    "    # using join() \n",
    "    # avoiding printing last comma \n",
    "    for gene in sym_to_fix:\n",
    "        print(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print_list(interactome_seed, 'interactorA', 'interactorB')\n",
    "#Print this list and put the list to https://amp.pharm.mssm.edu/Enrichr Enrichr. Than download KEGG HUMAN 2019 and Ontologies tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sym_to_fix=[]\n",
    "sym_to_fix.extend(unionint['interactorA'])\n",
    "sym_to_fix.extend(unionint['interactorB'])\n",
    "#remove duplicates\n",
    "sym_to_fix=list(set(sym_to_fix))\n",
    "# using join() \n",
    "# avoiding printing last comma \n",
    "#print(', '.join(sym_to_fix)) \n",
    "\n",
    "# deleted quotes to pass the list into site\n",
    "#https://www.genenames.org/tools/multi-symbol-checker/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "approvedsym = pd.read_csv(\"approved-symbols.csv\", sep = ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "approvedsym.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, let's open the downloaded kegg and go tables and save only the first 10 entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_seed = pd.read_table(\"enrichr/kegg_human/seed/KEGG_2019_Human_table.txt\")\n",
    "kegg_seed = kegg_seed[:10]\n",
    "kegg_seed.to_csv(\"enrichr/kegg_human/seed/KEGG_2019_Human_table.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_union = pd.read_table(\"enrichr/kegg_human/union/KEGG_2019_Human_table.txt\")\n",
    "kegg_union = kegg_union[:10]\n",
    "kegg_union.to_csv(\"enrichr/kegg_human/union/KEGG_2019_Human_table_top10.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_bp_seed = pd.read_table(\"enrichr/ontologies/seed/GO_Biological_Process_2018_table.txt\")\n",
    "go_bp_seed = go_bp_seed[:10]\n",
    "go_bp_seed.to_csv(\"enrichr/ontologies/seed/GO_Biological_Process_2018_table.txt\")\n",
    "\n",
    "go_mf_seed = pd.read_table(\"enrichr/ontologies/seed/GO_Molecular_Function_2018_table.txt\")\n",
    "go_mf_seed = go_mf_seed[:10]\n",
    "go_mf_seed.to_csv(\"enrichr/ontologies/seed/GO_Molecular_Function_2018_table.txt\")\n",
    "\n",
    "go_cc_seed = pd.read_csv(\"enrichr/ontologies/seed/GO_Cellular_Component_2018_table.txt\")\n",
    "go_cc_seed = go_cc_seed[:10]\n",
    "go_cc_seed.to_csv(\"enrichr/ontologies/seed/GO_Cellular_Component_2018_table.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_cc_seed = pd.read_csv(\"enrichr/ontologies/seed/GO_Cellular_Component_2018_table.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_cc_seed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_bp_union = pd.read_table(\"enrichr/ontologies/union/GO_Biological_Process_2018_table.txt\")\n",
    "go_bp_union = go_bp_union[:10]\n",
    "go_bp_union.to_csv(\"enrichr/ontologies/union/GO_Biological_Process_2018_table.txt\")\n",
    "\n",
    "go_mf_union = pd.read_table(\"enrichr/ontologies/union/GO_Molecular_Function_2018_table.txt\")\n",
    "go_mf_union = go_mf_union[:10]\n",
    "go_mf_union.to_csv(\"enrichr/ontologies/union/GO_Molecular_Function_2018_table.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_cc_union = pd.read_csv(\"enrichr/ontologies/union/GO_Cellular_Component_2018_table.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Term</th>\n",
       "      <th>Overlap</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Adjusted P-value</th>\n",
       "      <th>Old P-value</th>\n",
       "      <th>Old Adjusted P-value</th>\n",
       "      <th>Odds Ratio</th>\n",
       "      <th>Combined Score</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>focal adhesion (GO:0005925)</td>\n",
       "      <td>246/356</td>\n",
       "      <td>2.0307521217463597E-63</td>\n",
       "      <td>9.057154e-61</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0</td>\n",
       "      <td>2.582737</td>\n",
       "      <td>372.829565</td>\n",
       "      <td>RPL4;CYFIP1;RPL5;MDC1;NCKAP1;RPL30;TES;RPL3;RP...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nuclear body (GO:0016604)</td>\n",
       "      <td>312/618</td>\n",
       "      <td>2.8626122970109088E-37</td>\n",
       "      <td>6.383625e-35</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0</td>\n",
       "      <td>1.886953</td>\n",
       "      <td>158.775640</td>\n",
       "      <td>RB1;MDC1;EHMT2;GPATCH2;RAPH1;HNRNPU;TESK2;MKI6...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nuclear chromosome part (GO:0044454)</td>\n",
       "      <td>223/392</td>\n",
       "      <td>6.3003903982783755E-37</td>\n",
       "      <td>9.366580e-35</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0</td>\n",
       "      <td>2.126248</td>\n",
       "      <td>177.233449</td>\n",
       "      <td>SPI1;SMARCB1;TRRAP;EHMT2;CCAR2;ACTB;KAT5;ZMIZ2...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chromatin (GO:0000785)</td>\n",
       "      <td>182/296</td>\n",
       "      <td>1.7710370351351232E-36</td>\n",
       "      <td>1.974706e-34</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0</td>\n",
       "      <td>2.298131</td>\n",
       "      <td>189.185561</td>\n",
       "      <td>RB1;SMARCB1;SPI1;MRE11;KDM1A;EHMT2;DSCC1;JMJD1...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nucleolus (GO:0005730)</td>\n",
       "      <td>329/676</td>\n",
       "      <td>3.2856822475828046E-35</td>\n",
       "      <td>2.930829e-33</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0</td>\n",
       "      <td>1.819048</td>\n",
       "      <td>144.434085</td>\n",
       "      <td>RPL4;EIF4A2;RPL5;JPT1;RPL3;SMARCB1;POP1;EIF4A3...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cytosolic part (GO:0044445)</td>\n",
       "      <td>114/159</td>\n",
       "      <td>2.5530499276119735E-32</td>\n",
       "      <td>1.897767e-30</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0</td>\n",
       "      <td>2.679802</td>\n",
       "      <td>194.943389</td>\n",
       "      <td>RPL4;RPL5;RPL30;RPL3;RPL31;RPL10L;AHR;RPL8;RPL...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>cytosolic ribosome (GO:0022626)</td>\n",
       "      <td>89/124</td>\n",
       "      <td>1.3972500602057802E-25</td>\n",
       "      <td>8.902479e-24</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0</td>\n",
       "      <td>2.682646</td>\n",
       "      <td>153.528155</td>\n",
       "      <td>RPL4;RPL5;RPL30;RPL3;DDX3X;RPL31;RPL10L;RPLP0;...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cytoskeleton (GO:0005856)</td>\n",
       "      <td>244/520</td>\n",
       "      <td>1.6517829255469964E-23</td>\n",
       "      <td>9.208690e-22</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0</td>\n",
       "      <td>1.753806</td>\n",
       "      <td>92.000452</td>\n",
       "      <td>ZNF174;SLC4A1;ACTB;MYLK;RASSF1;BAIAP2L1;MPRIP;...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chromosome</td>\n",
       "      <td>telomeric region (GO:0000781)</td>\n",
       "      <td>86/124</td>\n",
       "      <td>4.106995e-23</td>\n",
       "      <td>2.035244e-21</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.592220</td>\n",
       "      <td>133.6205419871832</td>\n",
       "      <td>FEN1;MRE11;DCLRE1C;KDM1A;MCM7;PINX1;NAT10;BRCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nuclear speck (GO:0016607)</td>\n",
       "      <td>159/296</td>\n",
       "      <td>4.484949237649398E-23</td>\n",
       "      <td>2.000287e-21</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0</td>\n",
       "      <td>2.007708</td>\n",
       "      <td>103.314082</td>\n",
       "      <td>EHMT2;GPATCH2;FAM107A;JADE1;HNRNPU;NR3C1;SMC4;...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>cytoplasmic ribonucleoprotein granule (GO:0036...</td>\n",
       "      <td>106/170</td>\n",
       "      <td>1.7821255079836267E-22</td>\n",
       "      <td>7.225709e-21</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0</td>\n",
       "      <td>2.330515</td>\n",
       "      <td>116.710036</td>\n",
       "      <td>FMR1;HNRNPU;ACTB;RPL6;TIAL1;EDC4;ZFP36;HABP4;M...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 Term  \\\n",
       "0                         focal adhesion (GO:0005925)   \n",
       "1                           nuclear body (GO:0016604)   \n",
       "2                nuclear chromosome part (GO:0044454)   \n",
       "3                              chromatin (GO:0000785)   \n",
       "4                              nucleolus (GO:0005730)   \n",
       "5                         cytosolic part (GO:0044445)   \n",
       "6                     cytosolic ribosome (GO:0022626)   \n",
       "7                           cytoskeleton (GO:0005856)   \n",
       "8                                          chromosome   \n",
       "9                          nuclear speck (GO:0016607)   \n",
       "10  cytoplasmic ribonucleoprotein granule (GO:0036...   \n",
       "\n",
       "                           Overlap                 P-value  Adjusted P-value  \\\n",
       "0                          246/356  2.0307521217463597E-63      9.057154e-61   \n",
       "1                          312/618  2.8626122970109088E-37      6.383625e-35   \n",
       "2                          223/392  6.3003903982783755E-37      9.366580e-35   \n",
       "3                          182/296  1.7710370351351232E-36      1.974706e-34   \n",
       "4                          329/676  3.2856822475828046E-35      2.930829e-33   \n",
       "5                          114/159  2.5530499276119735E-32      1.897767e-30   \n",
       "6                           89/124  1.3972500602057802E-25      8.902479e-24   \n",
       "7                          244/520  1.6517829255469964E-23      9.208690e-22   \n",
       "8    telomeric region (GO:0000781)                  86/124      4.106995e-23   \n",
       "9                          159/296   4.484949237649398E-23      2.000287e-21   \n",
       "10                         106/170  1.7821255079836267E-22      7.225709e-21   \n",
       "\n",
       "     Old P-value  Old Adjusted P-value  Odds Ratio  Combined Score  \\\n",
       "0   0.000000e+00                     0    2.582737      372.829565   \n",
       "1   0.000000e+00                     0    1.886953      158.775640   \n",
       "2   0.000000e+00                     0    2.126248      177.233449   \n",
       "3   0.000000e+00                     0    2.298131      189.185561   \n",
       "4   0.000000e+00                     0    1.819048      144.434085   \n",
       "5   0.000000e+00                     0    2.679802      194.943389   \n",
       "6   0.000000e+00                     0    2.682646      153.528155   \n",
       "7   0.000000e+00                     0    1.753806       92.000452   \n",
       "8   2.035244e-21                     0    0.000000        2.592220   \n",
       "9   0.000000e+00                     0    2.007708      103.314082   \n",
       "10  0.000000e+00                     0    2.330515      116.710036   \n",
       "\n",
       "                                                Genes  \\\n",
       "0   RPL4;CYFIP1;RPL5;MDC1;NCKAP1;RPL30;TES;RPL3;RP...   \n",
       "1   RB1;MDC1;EHMT2;GPATCH2;RAPH1;HNRNPU;TESK2;MKI6...   \n",
       "2   SPI1;SMARCB1;TRRAP;EHMT2;CCAR2;ACTB;KAT5;ZMIZ2...   \n",
       "3   RB1;SMARCB1;SPI1;MRE11;KDM1A;EHMT2;DSCC1;JMJD1...   \n",
       "4   RPL4;EIF4A2;RPL5;JPT1;RPL3;SMARCB1;POP1;EIF4A3...   \n",
       "5   RPL4;RPL5;RPL30;RPL3;RPL31;RPL10L;AHR;RPL8;RPL...   \n",
       "6   RPL4;RPL5;RPL30;RPL3;DDX3X;RPL31;RPL10L;RPLP0;...   \n",
       "7   ZNF174;SLC4A1;ACTB;MYLK;RASSF1;BAIAP2L1;MPRIP;...   \n",
       "8                                   133.6205419871832   \n",
       "9   EHMT2;GPATCH2;FAM107A;JADE1;HNRNPU;NR3C1;SMC4;...   \n",
       "10  FMR1;HNRNPU;ACTB;RPL6;TIAL1;EDC4;ZFP36;HABP4;M...   \n",
       "\n",
       "                                           Unnamed: 9  \n",
       "0                                                 NaN  \n",
       "1                                                 NaN  \n",
       "2                                                 NaN  \n",
       "3                                                 NaN  \n",
       "4                                                 NaN  \n",
       "5                                                 NaN  \n",
       "6                                                 NaN  \n",
       "7                                                 NaN  \n",
       "8   FEN1;MRE11;DCLRE1C;KDM1A;MCM7;PINX1;NAT10;BRCA...  \n",
       "9                                                 NaN  \n",
       "10                                                NaN  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_cc_union"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Term', 'Overlap', 'P-value', 'Adjusted P-value', 'Old P-value',\n",
       "       'Old Adjusted P-value', 'Odds Ratio', 'Combined Score', 'Genes',\n",
       "       'Unnamed: 9'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_cc_union.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
